{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Specify your target directory\n",
    "# Traverse directory\n",
    "# parse files to dict structure\n",
    "# Compute derivative values\n",
    "# Print medians"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "from os.path import dirname, join\n",
    "from glob import glob \n",
    "from statistics import median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def print_median(name, label, key, contents):\n",
    "    print(name, label, \"{:0.2%}\".format(median([contents[d][key] for d in contents.keys()])))\n",
    "    \n",
    "def print_average(name, label, key, contents):\n",
    "    avg = 0\n",
    "    total_length = 0\n",
    "    genome_length = sum([v[\"Ref Chr Total Size (No N's)\"] for k,v in contents.items()])\n",
    "    for chr in contents.keys():\n",
    "        avg += contents[chr][key] * contents[chr][\"Ref Chr Total Size (No N's)\"] / genome_length\n",
    "    print(name, label, \"{:0.2%}\".format(avg))\n",
    "\n",
    "def summarize_stats(name, contents):\n",
    "#     print(\"Median Chromosome Value\")\n",
    "#     print_median(name, 'identity', 'Ref Identity within Alignment', contents)\n",
    "#     print_median(name, 'coverage', 'Alignment Coverage of Ref Chr', contents)\n",
    "#     print_median(name, 'query coverage', \"Alignment Coverage of Query Main Chr\", contents)\n",
    "    \n",
    "    print_average(name, 'identity', 'Ref Identity within Alignment', contents)\n",
    "    print_average(name, 'coverage', 'Alignment Coverage of Ref Chr', contents)\n",
    "#     print_average(name, 'query coverage', \"Alignment Coverage of Query Main Chr\", contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def number(string):\n",
    "    try:\n",
    "        return int(string)\n",
    "    except ValueError:\n",
    "        return float(string)\n",
    "    \n",
    "def collect_stats(folder, prefix=''):\n",
    "    contents = {}\n",
    "      # Specify you target directory\n",
    "    os.chdir(folder)\n",
    "    # Traverse directory\n",
    "    subdirs = glob(prefix + '*')\n",
    "    for d in subdirs:\n",
    "        stats_file = os.path.join(d, r'stats.txt')\n",
    "        if os.path.exists(stats_file):\n",
    "            with open(stats_file) as stat:\n",
    "                # parse files to dict structure\n",
    "                lines = [line.split('\\t') for line in stat.read().splitlines()]\n",
    "                v = {line[0]: number(line[1]) for line in lines if len(line) == 2}\n",
    "                # Compute derivative values\n",
    "                v[\"Total alignment Length\"] = v['Aligned Variance in bp'] + v['Shared seq bp']\n",
    "                v[\"Ref Chr Total Size (No N's)\"] = v['Ref unique bp'] + v[\"Total alignment Length\"]\n",
    "                v[\"Ref Identity within Alignment\"] = 1 - (v['Aligned Variance in bp'] / v['Total alignment Length'])\n",
    "                v[\"Alignment Coverage of Ref Chr\"] = v[\"Total alignment Length\"] / v[\"Ref Chr Total Size (No N's)\"]\n",
    "                v[\"Alignment Coverage of Query Main Chr\"] = v['Total alignment Length'] / (v['Total alignment Length'] + v[\"Query unique bp\"])\n",
    "                contents[d] = v\n",
    "    # Print medians\n",
    "    name = prefix or os.path.basename(folder)\n",
    "#     print(\"Averages Weighted by Chromosome Length\")\n",
    "    summarize_stats(name, contents)\n",
    "    return contents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Human Hg38 vs Chimpanzee PanTro6_ identity 98.65%\n",
      "Human Hg38 vs Chimpanzee PanTro6_ coverage 94.95%\n",
      "\n",
      "\n",
      "chr1\n",
      "Ref Number of Gaps (all)\t165609\n",
      "Ref Gaps larger than 10bp\t18763\n",
      "Query Number of Gaps (all)\t171053\n",
      "Query Gaps larger than 10bp\t21270\n",
      "Ref Gaps larger than 100bp\t4479\n",
      "Query Gaps larger than 100bp\t4832\n",
      "Ref Gaps larger than 1000bp\t1655\n",
      "Query Gaps larger than 1000bp\t1943\n",
      "Ref N to query bp\t18475471\n",
      "Ref unique bp\t8720120\n",
      "Shared seq bp\t220917707\n",
      "Aligned Variance in bp\t2830836\n",
      "Query unique bp\t8724657\n",
      "Query N to ref in bp\t2495044\n",
      "Total alignment Length\t223748543\n",
      "Ref Chr Total Size (No N's)\t232468663\n",
      "Ref Identity within Alignment\t0.9873481366088717\n",
      "Alignment Coverage of Ref Chr\t0.9624890516964001\n",
      "Alignment Coverage of Query Main Chr\t0.9624702675405165\n",
      "\n",
      "\n",
      "chr10\n",
      "Query Number of Gaps (all)\t98327\n",
      "Query Gaps larger than 10bp\t12245\n",
      "Ref Number of Gaps (all)\t95484\n",
      "Ref Gaps larger than 10bp\t11050\n",
      "Query Gaps larger than 100bp\t2354\n",
      "Query Gaps larger than 1000bp\t620\n",
      "Ref Gaps larger than 100bp\t2504\n",
      "Ref Gaps larger than 1000bp\t819\n",
      "Ref N to query bp\t534460\n",
      "Ref unique bp\t6082908\n",
      "Shared seq bp\t126144862\n",
      "Aligned Variance in bp\t1708138\n",
      "Query unique bp\t4145888\n",
      "Query N to ref in bp\t1197562\n",
      "Total alignment Length\t127853000\n",
      "Ref Chr Total Size (No N's)\t133935908\n",
      "Ref Identity within Alignment\t0.986639828553104\n",
      "Alignment Coverage of Ref Chr\t0.9545834415069632\n",
      "Alignment Coverage of Query Main Chr\t0.9685914929828803\n",
      "\n",
      "\n",
      "chr11\n",
      "Query Number of Gaps (all)\t94230\n",
      "Ref Number of Gaps (all)\t92132\n",
      "Query Gaps larger than 10bp\t10997\n",
      "Query Gaps larger than 100bp\t1996\n",
      "Query Gaps larger than 1000bp\t598\n",
      "Ref Gaps larger than 10bp\t9946\n",
      "Ref Gaps larger than 100bp\t1971\n",
      "Ref Gaps larger than 1000bp\t677\n",
      "Ref N to query bp\t552880\n",
      "Shared seq bp\t126357568\n",
      "Aligned Variance in bp\t1694144\n",
      "Ref unique bp\t7284509\n",
      "Query unique bp\t2765835\n",
      "Query N to ref in bp\t8258\n",
      "Total alignment Length\t128051712\n",
      "Ref Chr Total Size (No N's)\t135336221\n",
      "Ref Identity within Alignment\t0.986769844982627\n",
      "Alignment Coverage of Ref Chr\t0.9461747273111756\n",
      "Alignment Coverage of Query Main Chr\t0.9788573087981844\n",
      "\n",
      "\n",
      "chr12\n",
      "Ref Number of Gaps (all)\t96419\n",
      "Query Number of Gaps (all)\t98479\n",
      "Ref Gaps larger than 10bp\t9907\n",
      "Ref Gaps larger than 100bp\t1901\n",
      "Query Gaps larger than 10bp\t11165\n",
      "Ref Gaps larger than 1000bp\t586\n",
      "Query Gaps larger than 100bp\t1887\n",
      "Query Gaps larger than 1000bp\t465\n",
      "Ref N to query bp\t137493\n",
      "Ref unique bp\t6103971\n",
      "Shared seq bp\t126059664\n",
      "Aligned Variance in bp\t1618815\n",
      "Query unique bp\t1956826\n",
      "Query N to ref in bp\t5774\n",
      "Total alignment Length\t127678479\n",
      "Ref Chr Total Size (No N's)\t133782450\n",
      "Ref Identity within Alignment\t0.987321160052353\n",
      "Alignment Coverage of Ref Chr\t0.9543739033034602\n",
      "Alignment Coverage of Query Main Chr\t0.9849051460171286\n",
      "\n",
      "\n",
      "chr13\n",
      "Ref Number of Gaps (all)\t71882\n",
      "Query Number of Gaps (all)\t74740\n",
      "Query Gaps larger than 10bp\t8273\n",
      "Query Gaps larger than 100bp\t1586\n",
      "Ref Gaps larger than 10bp\t7416\n",
      "Ref Gaps larger than 100bp\t1638\n",
      "Ref Gaps larger than 1000bp\t506\n",
      "Query Gaps larger than 1000bp\t408\n",
      "Ref N to query bp\t16381203\n",
      "Ref unique bp\t3849601\n",
      "Shared seq bp\t93347690\n",
      "Aligned Variance in bp\t1250938\n",
      "Query unique bp\t5205275\n",
      "Query N to ref in bp\t402562\n",
      "Total alignment Length\t94598628\n",
      "Ref Chr Total Size (No N's)\t98448229\n",
      "Ref Identity within Alignment\t0.9867763621265205\n",
      "Alignment Coverage of Ref Chr\t0.9608972041538706\n",
      "Alignment Coverage of Query Main Chr\t0.9478449755617273\n",
      "\n",
      "\n",
      "chr14\n",
      "Query Number of Gaps (all)\t66865\n",
      "Query Gaps larger than 10bp\t7557\n",
      "Ref Number of Gaps (all)\t65255\n",
      "Query Gaps larger than 100bp\t1328\n",
      "Query Gaps larger than 1000bp\t440\n",
      "Ref Gaps larger than 10bp\t6801\n",
      "Ref Gaps larger than 100bp\t1370\n",
      "Ref Gaps larger than 1000bp\t462\n",
      "Ref N to query bp\t16475569\n",
      "Ref unique bp\t4245909\n",
      "Shared seq bp\t85325045\n",
      "Query unique bp\t2098847\n",
      "Aligned Variance in bp\t1134250\n",
      "Query N to ref in bp\t475733\n",
      "Total alignment Length\t86459295\n",
      "Ref Chr Total Size (No N's)\t90705204\n",
      "Ref Identity within Alignment\t0.9868811097754151\n",
      "Alignment Coverage of Ref Chr\t0.9531900176311825\n",
      "Alignment Coverage of Query Main Chr\t0.9762997850609829\n",
      "\n",
      "\n",
      "chr15\n",
      "Query Number of Gaps (all)\t66261\n",
      "Ref Number of Gaps (all)\t63744\n",
      "Query Gaps larger than 10bp\t7551\n",
      "Ref Gaps larger than 10bp\t6843\n",
      "Query Gaps larger than 100bp\t1404\n",
      "Query Gaps larger than 1000bp\t478\n",
      "Ref Gaps larger than 100bp\t1455\n",
      "Ref Gaps larger than 1000bp\t578\n",
      "Ref N to query bp\t17349864\n",
      "Ref unique bp\t4447223\n",
      "Shared seq bp\t79377739\n",
      "Aligned Variance in bp\t1110620\n",
      "Query unique bp\t9776198\n",
      "Query N to ref in bp\t1518735\n",
      "Total alignment Length\t80488359\n",
      "Ref Chr Total Size (No N's)\t84935582\n",
      "Ref Identity within Alignment\t0.9862014828753062\n",
      "Alignment Coverage of Ref Chr\t0.9476400479601117\n",
      "Alignment Coverage of Query Main Chr\t0.8916939458308093\n",
      "\n",
      "\n",
      "chr16\n",
      "Query Number of Gaps (all)\t70208\n",
      "Query Gaps larger than 10bp\t10578\n",
      "Ref Number of Gaps (all)\t66458\n",
      "Ref Gaps larger than 10bp\t9996\n",
      "Ref Gaps larger than 100bp\t2892\n",
      "Query Gaps larger than 100bp\t2723\n",
      "Ref Gaps larger than 1000bp\t1127\n",
      "Query Gaps larger than 1000bp\t969\n",
      "Ref N to query bp\t8533476\n",
      "Ref unique bp\t4824479\n",
      "Shared seq bp\t76802346\n",
      "Query unique bp\t8134807\n",
      "Aligned Variance in bp\t1219533\n",
      "Query N to ref in bp\t355075\n",
      "Total alignment Length\t78021879\n",
      "Ref Chr Total Size (No N's)\t82846358\n",
      "Ref Identity within Alignment\t0.9843693459369263\n",
      "Alignment Coverage of Ref Chr\t0.9417659494458405\n",
      "Alignment Coverage of Query Main Chr\t0.9055812453139156\n",
      "\n",
      "\n",
      "chr17\n",
      "Query Number of Gaps (all)\t64542\n",
      "Ref Number of Gaps (all)\t64312\n",
      "Query Gaps larger than 10bp\t8707\n",
      "Ref Gaps larger than 10bp\t8164\n",
      "Query Gaps larger than 100bp\t1727\n",
      "Ref Gaps larger than 100bp\t1826\n",
      "Ref Gaps larger than 1000bp\t598\n",
      "Query Gaps larger than 1000bp\t550\n",
      "Ref N to query bp\t337994\n",
      "Shared seq bp\t75807242\n",
      "Aligned Variance in bp\t1004791\n",
      "Query unique bp\t33191258\n",
      "Ref unique bp\t6846475\n",
      "Query N to ref in bp\t853093\n",
      "Total alignment Length\t76812033\n",
      "Ref Chr Total Size (No N's)\t83658508\n",
      "Ref Identity within Alignment\t0.9869188334072605\n",
      "Alignment Coverage of Ref Chr\t0.9181616411327823\n",
      "Alignment Coverage of Query Main Chr\t0.6982703181125736\n",
      "\n",
      "\n",
      "chr18\n",
      "Ref Number of Gaps (all)\t55079\n",
      "Query Number of Gaps (all)\t56834\n",
      "Query Gaps larger than 10bp\t6424\n",
      "Ref Gaps larger than 10bp\t5861\n",
      "Query Gaps larger than 100bp\t1148\n",
      "Ref Gaps larger than 100bp\t1129\n",
      "Query Gaps larger than 1000bp\t283\n",
      "Ref Gaps larger than 1000bp\t323\n",
      "Ref N to query bp\t283680\n",
      "Ref unique bp\t6855876\n",
      "Shared seq bp\t72457276\n",
      "Aligned Variance in bp\t978685\n",
      "Query unique bp\t1233057\n",
      "Query N to ref in bp\t421540\n",
      "Total alignment Length\t73435961\n",
      "Ref Chr Total Size (No N's)\t80291837\n",
      "Ref Identity within Alignment\t0.9866729462422368\n",
      "Alignment Coverage of Ref Chr\t0.9146130384337825\n",
      "Alignment Coverage of Query Main Chr\t0.9834863637820976\n",
      "\n",
      "\n",
      "chr19\n",
      "Ref Number of Gaps (all)\t53722\n",
      "Ref Gaps larger than 10bp\t7683\n",
      "Query Number of Gaps (all)\t54235\n",
      "Query Gaps larger than 10bp\t8220\n",
      "Query Gaps larger than 100bp\t1978\n",
      "Query Gaps larger than 1000bp\t608\n",
      "Ref Gaps larger than 100bp\t1990\n",
      "Ref Gaps larger than 1000bp\t667\n",
      "Ref N to query bp\t176858\n",
      "Shared seq bp\t52358698\n",
      "Aligned Variance in bp\t831326\n",
      "Ref unique bp\t5993312\n",
      "Query unique bp\t2981850\n",
      "Query N to ref in bp\t509942\n",
      "Total alignment Length\t53190024\n",
      "Ref Chr Total Size (No N's)\t59183336\n",
      "Ref Identity within Alignment\t0.9843706406299045\n",
      "Alignment Coverage of Ref Chr\t0.898733116362349\n",
      "Alignment Coverage of Query Main Chr\t0.946915604061919\n",
      "\n",
      "\n",
      "chr2\n",
      "Ref Number of Gaps (all)\t170672\n",
      "Query Number of Gaps (all)\t178311\n",
      "Query Gaps larger than 10bp\t20429\n",
      "Ref Gaps larger than 10bp\t18015\n",
      "Ref Gaps larger than 100bp\t3618\n",
      "Query Gaps larger than 100bp\t3701\n",
      "Query Gaps larger than 1000bp\t1068\n",
      "Ref Gaps larger than 1000bp\t1187\n",
      "Ref N to query bp\t1645301\n",
      "Ref unique bp\t8260400\n",
      "Shared seq bp\t230534885\n",
      "Aligned Variance in bp\t3039716\n",
      "Query unique bp\t4682573\n",
      "Query N to ref in bp\t512888\n",
      "Total alignment Length\t233574601\n",
      "Ref Chr Total Size (No N's)\t241835001\n",
      "Ref Identity within Alignment\t0.9869861021404464\n",
      "Alignment Coverage of Ref Chr\t0.9658428268619397\n",
      "Alignment Coverage of Query Main Chr\t0.9803465603096594\n",
      "\n",
      "\n",
      "chr20\n",
      "Query Number of Gaps (all)\t47098\n",
      "Ref Number of Gaps (all)\t45779\n",
      "Query Gaps larger than 10bp\t6115\n",
      "Ref Gaps larger than 10bp\t5898\n",
      "Ref Gaps larger than 100bp\t1291\n",
      "Query Gaps larger than 100bp\t1104\n",
      "Ref Gaps larger than 1000bp\t405\n",
      "Query Gaps larger than 1000bp\t335\n",
      "Ref N to query bp\t500575\n",
      "Shared seq bp\t58788482\n",
      "Aligned Variance in bp\t834315\n",
      "Query unique bp\t1090382\n",
      "Ref unique bp\t4674499\n",
      "Query N to ref in bp\t65675\n",
      "Total alignment Length\t59622797\n",
      "Ref Chr Total Size (No N's)\t64297296\n",
      "Ref Identity within Alignment\t0.9860067785816892\n",
      "Alignment Coverage of Ref Chr\t0.9272986689829071\n",
      "Alignment Coverage of Query Main Chr\t0.9820404396877324\n",
      "\n",
      "\n",
      "chr21\n",
      "Query Number of Gaps (all)\t35036\n",
      "Ref Number of Gaps (all)\t33180\n",
      "Ref Gaps larger than 10bp\t4281\n",
      "Query Gaps larger than 10bp\t4377\n",
      "Query Gaps larger than 100bp\t909\n",
      "Query Gaps larger than 1000bp\t323\n",
      "Ref Gaps larger than 100bp\t1027\n",
      "Ref Gaps larger than 1000bp\t326\n",
      "Ref N to query bp\t6621364\n",
      "Shared seq bp\t35871105\n",
      "Aligned Variance in bp\t588978\n",
      "Query unique bp\t910924\n",
      "Ref unique bp\t3798511\n",
      "Query N to ref in bp\t180762\n",
      "Total alignment Length\t36460083\n",
      "Ref Chr Total Size (No N's)\t40258594\n",
      "Ref Identity within Alignment\t0.9838459501038437\n",
      "Alignment Coverage of Ref Chr\t0.9056472016881663\n",
      "Alignment Coverage of Query Main Chr\t0.9756248473582743\n",
      "\n",
      "\n",
      "chr22\n",
      "Query Number of Gaps (all)\t35142\n",
      "Ref Number of Gaps (all)\t33888\n",
      "Query Gaps larger than 10bp\t5098\n",
      "Query Gaps larger than 100bp\t1200\n",
      "Query Gaps larger than 1000bp\t443\n",
      "Ref Gaps larger than 10bp\t5000\n",
      "Ref Gaps larger than 100bp\t1311\n",
      "Ref Gaps larger than 1000bp\t481\n",
      "Ref N to query bp\t11659387\n",
      "Shared seq bp\t35177263\n",
      "Aligned Variance in bp\t598637\n",
      "Ref unique bp\t3699846\n",
      "Query unique bp\t5071519\n",
      "Query N to ref in bp\t278066\n",
      "Total alignment Length\t35775900\n",
      "Ref Chr Total Size (No N's)\t39475746\n",
      "Ref Identity within Alignment\t0.983267031716882\n",
      "Alignment Coverage of Ref Chr\t0.9062754634199947\n",
      "Alignment Coverage of Query Main Chr\t0.8758423635040442\n",
      "\n",
      "\n",
      "chr3\n",
      "Ref Number of Gaps (all)\t137201\n",
      "Query Number of Gaps (all)\t142805\n",
      "Query Gaps larger than 10bp\t15560\n",
      "Ref Gaps larger than 10bp\t13383\n",
      "Ref Gaps larger than 100bp\t2374\n",
      "Query Gaps larger than 100bp\t2525\n",
      "Query Gaps larger than 1000bp\t662\n",
      "Ref Gaps larger than 1000bp\t718\n",
      "Ref N to query bp\t195424\n",
      "Ref unique bp\t6981350\n",
      "Shared seq bp\t189497504\n",
      "Aligned Variance in bp\t2455454\n",
      "Query unique bp\t2722691\n",
      "Query N to ref in bp\t454701\n",
      "Total alignment Length\t191952958\n",
      "Ref Chr Total Size (No N's)\t198934308\n",
      "Ref Identity within Alignment\t0.9872080429205994\n",
      "Alignment Coverage of Ref Chr\t0.9649062543802148\n",
      "Alignment Coverage of Query Main Chr\t0.9860142189637698\n",
      "\n",
      "\n",
      "chr4\n",
      "Query Number of Gaps (all)\t145896\n",
      "Ref Number of Gaps (all)\t139495\n",
      "Query Gaps larger than 10bp\t15261\n",
      "Ref Gaps larger than 10bp\t13788\n",
      "Ref Gaps larger than 100bp\t2773\n",
      "Query Gaps larger than 100bp\t2840\n",
      "Ref Gaps larger than 1000bp\t912\n",
      "Query Gaps larger than 1000bp\t832\n",
      "Ref N to query bp\t461888\n",
      "Ref unique bp\t6721519\n",
      "Shared seq bp\t181435990\n",
      "Aligned Variance in bp\t2501290\n",
      "Query unique bp\t3241496\n",
      "Query N to ref in bp\t880917\n",
      "Total alignment Length\t183937280\n",
      "Ref Chr Total Size (No N's)\t190658799\n",
      "Ref Identity within Alignment\t0.9864013972588918\n",
      "Alignment Coverage of Ref Chr\t0.9647458232441714\n",
      "Alignment Coverage of Query Main Chr\t0.9826823528325669\n",
      "\n",
      "\n",
      "chr5\n",
      "Ref Number of Gaps (all)\t127304\n",
      "Query Number of Gaps (all)\t131709\n",
      "Query Gaps larger than 10bp\t14357\n",
      "Ref Gaps larger than 10bp\t12607\n",
      "Ref Gaps larger than 100bp\t2396\n",
      "Ref Gaps larger than 1000bp\t828\n",
      "Query Gaps larger than 100bp\t2518\n",
      "Query Gaps larger than 1000bp\t725\n",
      "Ref N to query bp\t272881\n",
      "Ref unique bp\t8377245\n",
      "Shared seq bp\t171933266\n",
      "Aligned Variance in bp\t2254345\n",
      "Query unique bp\t30017302\n",
      "Query N to ref in bp\t712412\n",
      "Total alignment Length\t174187611\n",
      "Ref Chr Total Size (No N's)\t182564856\n",
      "Ref Identity within Alignment\t0.9870579486849957\n",
      "Alignment Coverage of Ref Chr\t0.9541135945682777\n",
      "Alignment Coverage of Query Main Chr\t0.853004016607573\n",
      "\n",
      "\n",
      "chr6\n",
      "Ref Number of Gaps (all)\t123905\n",
      "Ref Gaps larger than 10bp\t12484\n",
      "Query Number of Gaps (all)\t125787\n",
      "Ref Gaps larger than 100bp\t2308\n",
      "Ref Gaps larger than 1000bp\t688\n",
      "Query Gaps larger than 10bp\t13711\n",
      "Query Gaps larger than 100bp\t2331\n",
      "Query Gaps larger than 1000bp\t633\n",
      "Ref N to query bp\t727457\n",
      "Shared seq bp\t163553827\n",
      "Aligned Variance in bp\t2128352\n",
      "Ref unique bp\t5024550\n",
      "Query unique bp\t3922784\n",
      "Query N to ref in bp\t424267\n",
      "Total alignment Length\t165682179\n",
      "Ref Chr Total Size (No N's)\t170706729\n",
      "Ref Identity within Alignment\t0.9871540076739334\n",
      "Alignment Coverage of Ref Chr\t0.9705661866439957\n",
      "Alignment Coverage of Query Main Chr\t0.9768710541801775\n",
      "\n",
      "\n",
      "chr7\n",
      "Query Number of Gaps (all)\t124269\n",
      "Ref Number of Gaps (all)\t119253\n",
      "Ref Gaps larger than 10bp\t14143\n",
      "Query Gaps larger than 10bp\t15879\n",
      "Query Gaps larger than 100bp\t3533\n",
      "Query Gaps larger than 1000bp\t1163\n",
      "Ref Gaps larger than 100bp\t3573\n",
      "Ref Gaps larger than 1000bp\t1277\n",
      "Ref N to query bp\t376941\n",
      "Ref unique bp\t8153166\n",
      "Shared seq bp\t149955018\n",
      "Aligned Variance in bp\t2064825\n",
      "Query unique bp\t6402326\n",
      "Query N to ref in bp\t2041529\n",
      "Total alignment Length\t152019843\n",
      "Ref Chr Total Size (No N's)\t160173009\n",
      "Ref Identity within Alignment\t0.9864173981550554\n",
      "Alignment Coverage of Ref Chr\t0.9490977534173688\n",
      "Alignment Coverage of Query Main Chr\t0.9595869312962128\n",
      "\n",
      "\n",
      "chr8\n",
      "Query Number of Gaps (all)\t107404\n",
      "Ref Number of Gaps (all)\t103879\n",
      "Query Gaps larger than 10bp\t12314\n",
      "Query Gaps larger than 100bp\t2183\n",
      "Ref Gaps larger than 10bp\t11382\n",
      "Query Gaps larger than 1000bp\t565\n",
      "Ref Gaps larger than 100bp\t2359\n",
      "Ref Gaps larger than 1000bp\t739\n",
      "Ref N to query bp\t370500\n",
      "Ref unique bp\t6336318\n",
      "Shared seq bp\t137307451\n",
      "Aligned Variance in bp\t1984435\n",
      "Query unique bp\t2323770\n",
      "Query N to ref in bp\t9448\n",
      "Total alignment Length\t139291886\n",
      "Ref Chr Total Size (No N's)\t145628204\n",
      "Ref Identity within Alignment\t0.9857534056219183\n",
      "Alignment Coverage of Ref Chr\t0.956489760733436\n",
      "Alignment Coverage of Query Main Chr\t0.9835910091748613\n",
      "\n",
      "\n",
      "chr9\n",
      "Query Number of Gaps (all)\t95198\n",
      "Ref Number of Gaps (all)\t90319\n",
      "Ref Gaps larger than 10bp\t10275\n",
      "Query Gaps larger than 10bp\t10949\n",
      "Ref Gaps larger than 100bp\t2207\n",
      "Query Gaps larger than 100bp\t2123\n",
      "Query Gaps larger than 1000bp\t713\n",
      "Ref Gaps larger than 1000bp\t845\n",
      "Ref N to query bp\t16604167\n",
      "Ref unique bp\t6501557\n",
      "Shared seq bp\t114437369\n",
      "Aligned Variance in bp\t1639734\n",
      "Query unique bp\t2009904\n",
      "Query N to ref in bp\t240403\n",
      "Total alignment Length\t116077103\n",
      "Ref Chr Total Size (No N's)\t122578660\n",
      "Ref Identity within Alignment\t0.9858737515184196\n",
      "Alignment Coverage of Ref Chr\t0.9469601234015774\n",
      "Alignment Coverage of Query Main Chr\t0.9829794653022241\n",
      "\n",
      "\n",
      "chrX\n",
      "Query Number of Gaps (all)\t97074\n",
      "Ref Number of Gaps (all)\t89016\n",
      "Ref Gaps larger than 10bp\t10685\n",
      "Query Gaps larger than 10bp\t11746\n",
      "Ref Gaps larger than 100bp\t2703\n",
      "Ref Gaps larger than 1000bp\t1077\n",
      "Query Gaps larger than 100bp\t2534\n",
      "Query Gaps larger than 1000bp\t950\n",
      "Ref N to query bp\t1147971\n",
      "Ref unique bp\t10676764\n",
      "Shared seq bp\t143853741\n",
      "Aligned Variance in bp\t1536778\n",
      "Query unique bp\t2966721\n",
      "Query N to ref in bp\t1094639\n",
      "Total alignment Length\t145390519\n",
      "Ref Chr Total Size (No N's)\t156067283\n",
      "Ref Identity within Alignment\t0.9894299985269328\n",
      "Alignment Coverage of Ref Chr\t0.9315887110048555\n",
      "Alignment Coverage of Query Main Chr\t0.9800028566182547\n",
      "\n",
      "\n",
      "chrY\n",
      "Query Number of Gaps (all)\t35425\n",
      "Ref Number of Gaps (all)\t35422\n",
      "Ref Gaps larger than 10bp\t7320\n",
      "Query Gaps larger than 10bp\t7327\n",
      "Query Gaps larger than 100bp\t3123\n",
      "Ref Gaps larger than 100bp\t3226\n",
      "Ref Gaps larger than 1000bp\t1430\n",
      "Query Gaps larger than 1000bp\t1371\n",
      "Ref N to query bp\t30813205\n",
      "Ref unique bp\t4713798\n",
      "Shared seq bp\t22309259\n",
      "Aligned Variance in bp\t758461\n",
      "Query unique bp\t5105610\n",
      "Total alignment Length\t23067720\n",
      "Ref Chr Total Size (No N's)\t27781518\n",
      "Ref Identity within Alignment\t0.9671202442200616\n",
      "Alignment Coverage of Ref Chr\t0.8303261182488301\n",
      "Alignment Coverage of Query Main Chr\t0.8187786108351409\n"
     ]
    }
   ],
   "source": [
    "root = r'E:\\Projects\\FluentDNA-2.4.1\\www-data\\dnadata'\n",
    "# folder = os.path.join(root, )\n",
    "genome = collect_stats(root, prefix='Human Hg38 vs Chimpanzee PanTro6_')\n",
    "for name, stats in genome.items():\n",
    "    print(f\"\\n\\n{name[33:]}\")\n",
    "    for k,v in stats.items():\n",
    "        print(f'{k}\\t{v}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Ref Number of Gaps (all)', 'Ref Gaps larger than 10bp', 'Query Number of Gaps (all)', 'Query Gaps larger than 10bp', 'Ref Gaps larger than 100bp', 'Query Gaps larger than 100bp', 'Ref Gaps larger than 1000bp', 'Query Gaps larger than 1000bp', 'Ref N to query bp', 'Ref unique bp', 'Shared seq bp', 'Aligned Variance in bp', 'Query unique bp', 'Query N to ref in bp', 'Total alignment Length', \"Ref Chr Total Size (No N's)\", 'Ref Identity within Alignment', 'Alignment Coverage of Ref Chr', 'Alignment Coverage of Query Main Chr']\n",
      "chr1\tchr2\tchr3\tchr4\tchr5\tchr6\tchr7\tchr8\tchr9\tchr10\tchr11\tchr12\tchr13\tchr14\tchr15\tchr16\tchr17\tchr18\tchr19\tchr20\tchr21\tchr22\tchrX\tchrY\n",
      "Ref Number of Gaps (all)\t165609\t170672\t137201\t139495\t127304\t123905\t119253\t103879\t90319\t95484\t92132\t96419\t71882\t65255\t63744\t66458\t64312\t55079\t53722\t45779\t33180\t33888\t89016\t35422\t\n",
      "Ref Gaps larger than 10bp\t18763\t18015\t13383\t13788\t12607\t12484\t14143\t11382\t10275\t11050\t9946\t9907\t7416\t6801\t6843\t9996\t8164\t5861\t7683\t5898\t4281\t5000\t10685\t7320\t\n",
      "Query Number of Gaps (all)\t171053\t178311\t142805\t145896\t131709\t125787\t124269\t107404\t95198\t98327\t94230\t98479\t74740\t66865\t66261\t70208\t64542\t56834\t54235\t47098\t35036\t35142\t97074\t35425\t\n",
      "Query Gaps larger than 10bp\t21270\t20429\t15560\t15261\t14357\t13711\t15879\t12314\t10949\t12245\t10997\t11165\t8273\t7557\t7551\t10578\t8707\t6424\t8220\t6115\t4377\t5098\t11746\t7327\t\n",
      "Ref Gaps larger than 100bp\t4479\t3618\t2374\t2773\t2396\t2308\t3573\t2359\t2207\t2504\t1971\t1901\t1638\t1370\t1455\t2892\t1826\t1129\t1990\t1291\t1027\t1311\t2703\t3226\t\n",
      "Query Gaps larger than 100bp\t4832\t3701\t2525\t2840\t2518\t2331\t3533\t2183\t2123\t2354\t1996\t1887\t1586\t1328\t1404\t2723\t1727\t1148\t1978\t1104\t909\t1200\t2534\t3123\t\n",
      "Ref Gaps larger than 1000bp\t1655\t1187\t718\t912\t828\t688\t1277\t739\t845\t819\t677\t586\t506\t462\t578\t1127\t598\t323\t667\t405\t326\t481\t1077\t1430\t\n",
      "Query Gaps larger than 1000bp\t1943\t1068\t662\t832\t725\t633\t1163\t565\t713\t620\t598\t465\t408\t440\t478\t969\t550\t283\t608\t335\t323\t443\t950\t1371\t\n",
      "Ref N to query bp\t18475471\t1645301\t195424\t461888\t272881\t727457\t376941\t370500\t16604167\t534460\t552880\t137493\t16381203\t16475569\t17349864\t8533476\t337994\t283680\t176858\t500575\t6621364\t11659387\t1147971\t30813205\t\n",
      "Ref unique bp\t8720120\t8260400\t6981350\t6721519\t8377245\t5024550\t8153166\t6336318\t6501557\t6082908\t7284509\t6103971\t3849601\t4245909\t4447223\t4824479\t6846475\t6855876\t5993312\t4674499\t3798511\t3699846\t10676764\t4713798\t\n",
      "Shared seq bp\t220917707\t230534885\t189497504\t181435990\t171933266\t163553827\t149955018\t137307451\t114437369\t126144862\t126357568\t126059664\t93347690\t85325045\t79377739\t76802346\t75807242\t72457276\t52358698\t58788482\t35871105\t35177263\t143853741\t22309259\t\n",
      "Aligned Variance in bp\t2830836\t3039716\t2455454\t2501290\t2254345\t2128352\t2064825\t1984435\t1639734\t1708138\t1694144\t1618815\t1250938\t1134250\t1110620\t1219533\t1004791\t978685\t831326\t834315\t588978\t598637\t1536778\t758461\t\n",
      "Query unique bp\t8724657\t4682573\t2722691\t3241496\t30017302\t3922784\t6402326\t2323770\t2009904\t4145888\t2765835\t1956826\t5205275\t2098847\t9776198\t8134807\t33191258\t1233057\t2981850\t1090382\t910924\t5071519\t2966721\t5105610\t\n",
      "Query N to ref in bp\t2495044\t512888\t454701\t880917\t712412\t424267\t2041529\t9448\t240403\t1197562\t8258\t5774\t402562\t475733\t1518735\t355075\t853093\t421540\t509942\t65675\t180762\t278066\t1094639\t0\t\n",
      "Total alignment Length\t223748543\t233574601\t191952958\t183937280\t174187611\t165682179\t152019843\t139291886\t116077103\t127853000\t128051712\t127678479\t94598628\t86459295\t80488359\t78021879\t76812033\t73435961\t53190024\t59622797\t36460083\t35775900\t145390519\t23067720\t\n",
      "Ref Chr Total Size (No N's)\t232468663\t241835001\t198934308\t190658799\t182564856\t170706729\t160173009\t145628204\t122578660\t133935908\t135336221\t133782450\t98448229\t90705204\t84935582\t82846358\t83658508\t80291837\t59183336\t64297296\t40258594\t39475746\t156067283\t27781518\t\n",
      "Ref Identity within Alignment\t0.9873481366088717\t0.9869861021404464\t0.9872080429205994\t0.9864013972588918\t0.9870579486849957\t0.9871540076739334\t0.9864173981550554\t0.9857534056219183\t0.9858737515184196\t0.986639828553104\t0.986769844982627\t0.987321160052353\t0.9867763621265205\t0.9868811097754151\t0.9862014828753062\t0.9843693459369263\t0.9869188334072605\t0.9866729462422368\t0.9843706406299045\t0.9860067785816892\t0.9838459501038437\t0.983267031716882\t0.9894299985269328\t0.9671202442200616\t\n",
      "Alignment Coverage of Ref Chr\t0.9624890516964001\t0.9658428268619397\t0.9649062543802148\t0.9647458232441714\t0.9541135945682777\t0.9705661866439957\t0.9490977534173688\t0.956489760733436\t0.9469601234015774\t0.9545834415069632\t0.9461747273111756\t0.9543739033034602\t0.9608972041538706\t0.9531900176311825\t0.9476400479601117\t0.9417659494458405\t0.9181616411327823\t0.9146130384337825\t0.898733116362349\t0.9272986689829071\t0.9056472016881663\t0.9062754634199947\t0.9315887110048555\t0.8303261182488301\t\n",
      "Alignment Coverage of Query Main Chr\t0.9624702675405165\t0.9803465603096594\t0.9860142189637698\t0.9826823528325669\t0.853004016607573\t0.9768710541801775\t0.9595869312962128\t0.9835910091748613\t0.9829794653022241\t0.9685914929828803\t0.9788573087981844\t0.9849051460171286\t0.9478449755617273\t0.9762997850609829\t0.8916939458308093\t0.9055812453139156\t0.6982703181125736\t0.9834863637820976\t0.946915604061919\t0.9820404396877324\t0.9756248473582743\t0.8758423635040442\t0.9800028566182547\t0.8187786108351409\t\n"
     ]
    }
   ],
   "source": [
    "chr_order = 'chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 chrX chrY'.split()\n",
    "prefix = 'Human Hg38 vs Chimpanzee PanTro6_'\n",
    "labels = list(genome['Human Hg38 vs Chimpanzee PanTro6_chr1'].keys())\n",
    "print(labels)\n",
    "print('\\t'.join(chr_order))\n",
    "for label in labels:\n",
    "    if label:\n",
    "        print(label, end='\\t')\n",
    "        for c in chr_order:\n",
    "            \n",
    "            v = genome[prefix + c][label] if label in genome[prefix + c] else 0\n",
    "            print(v, end='\\t')\n",
    "        print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hg38toPantro6 identity 98.67%\n",
      "hg38toPantro6 coverage 91.46%\n",
      "\n"
     ]
    }
   ],
   "source": [
    "all_results = {}\n",
    "for comparison in ['hg38toPantro6']:\n",
    "    all_results[comparison] = collect_stats(comparison)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3370400\t2076300\t3157700\t2077200\t4186600\t1549700\t4162600\t2802300\t2600400\t2862200\t4463500\t3252100\t2049700\t2306200\t2779400\t2481000\t4171000\t5434500\t3408800\t3163400\t2244000\t2516800\t4574000\t662700\t"
     ]
    }
   ],
   "source": [
    "centromeres = {}\n",
    "lines = open(r\"E:\\Google Drive\\Research\\Thesis - Genome Symmetry\\Genome Symmetry Thesis Manuscript\\Actively Used Figures\\3\\Unique Human Centromere Locations.txt\", 'r').readlines()\n",
    "for line_i in range(0, len(lines), 2):\n",
    "    label1, start = lines[line_i].split('\\t')\n",
    "    label2, end = lines[line_i+1].split('\\t')\n",
    "    start, end = int(start), int(end)\n",
    "    assert label1 == label2 and start < end, (label1, label2, start, end)\n",
    "    centromeres[label1] = end - start\n",
    "for c in chr_order:\n",
    "    print(centromeres[c], end='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stats for Syntenic versus Translocation from Chain file\n",
    "We want to know how much of the alignment is a syntenic transfer 1:1 and how many bp have been rearranged between the two genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from DDV.ChainFiles import chain_file_to_list, fetch_all_chains, Chain, ChainEntry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28612"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig_names = 'chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 chrX chrY'.split()\n",
    "chain_stats = {name: {} for name in contig_names}\n",
    "chain_path = r\"D:\\Genomes\\Human\\alignments\\hg38ToPanTro6.over.chain\"\n",
    "headers = chain_file_to_list(chain_path, contig_names)  # just read the headers and discard the rest\n",
    "len(headers)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Chains are already sorted by highest score first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from DNASkittleUtils.DDVUtils import pp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('11,079,226,391', '242,193,529', '128,755,405')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pp(headers[0].score), pp(headers[0].tSize), pp(headers[0].qSize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'124,837,690'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pp(sum([e.size for e in headers[0].entries]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The actual coverage tends to be slightly less than the list \"query Size\" which includes gaps. It appears chain Score is not bp at all.  Likely this is a z-score."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sizes include difference already summed in.  I only know the % identity or total number of shared bp.  I don't know how that divides between my 3 categories, though if I assume it was evenly distributed, you could estimate shared bp in each category.  If I went through the unique sequence after having collapsed all ref gaps, then I'd have the exact number, since differences need not be attributed to a particular category.  We can switch to the translocations + non_syntenic + primary_chain that I've already calculated when we want to include differences.  \"No Coverage\" is already known.  \n",
    "**Conclusion**: These stats are all coverage, not shared bp or identity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 chr2B None : 478\n",
      "chr2 chr2A None : 798\n",
      "chr1 chr1 None : 1168\n",
      "chr1 None None : 1710\n",
      "chr2 chr2 None : 0\n",
      "chr2 None None : 1610\n",
      "chr3 chr3 None : 473\n",
      "chr3 None None : 661\n",
      "chr4 chr4 None : 859\n",
      "chr4 None None : 1126\n",
      "chr5 chr5 None : 811\n",
      "chr5 None None : 1206\n",
      "chr6 chr6 None : 797\n",
      "chr6 None None : 966\n",
      "chr7 chr7 None : 1262\n",
      "chr7 None None : 1643\n",
      "chr8 chr8 None : 927\n",
      "chr8 None None : 1135\n",
      "chr9 chr9 None : 588\n",
      "chr9 None None : 996\n",
      "chr10 chr10 None : 929\n",
      "chr10 None None : 1209\n",
      "chr11 chr11 None : 554\n",
      "chr11 None None : 701\n",
      "chr12 chr12 None : 613\n",
      "chr12 None None : 719\n",
      "chr13 chr13 None : 529\n",
      "chr13 None None : 701\n",
      "chr14 chr14 None : 477\n",
      "chr14 None None : 631\n",
      "chr15 chr15 None : 263\n",
      "chr15 None None : 461\n",
      "chr16 chr16 None : 752\n",
      "chr16 None None : 1105\n",
      "chr17 chr17 None : 748\n",
      "chr17 None None : 1003\n",
      "chr18 chr18 None : 586\n",
      "chr18 None None : 677\n",
      "chr19 chr19 None : 818\n",
      "chr19 None None : 986\n",
      "chr20 chr20 None : 505\n",
      "chr20 None None : 616\n",
      "chr21 chr21 None : 298\n",
      "chr21 None None : 541\n",
      "chr22 chr22 None : 530\n",
      "chr22 None None : 777\n",
      "chrX chrX None : 590\n",
      "chrX None None : 1488\n",
      "chrY chrY None : 313\n",
      "chrY None None : 1592\n"
     ]
    }
   ],
   "source": [
    "chr2_chains = fetch_all_chains('chr2', 'chr2B', None, headers) +\\\n",
    "            fetch_all_chains('chr2', 'chr2A', None, headers)  # special case chr2\n",
    "chr2_chains.sort(key=lambda chain: -chain.score)\n",
    "\n",
    "for chromosome in contig_names:\n",
    "    relevant = fetch_all_chains(chromosome, chromosome, None, headers)\n",
    "    if chromosome == 'chr2':\n",
    "        relevant = chr2_chains\n",
    "    chain_stats[chromosome]['Primary Chain'] = sum([e.size for e in relevant[0].entries])\n",
    "    non_syntenic = sum([e.size for chain in relevant[1:] for e in chain.entries]) if len(relevant) > 1 else 0\n",
    "    chain_stats[chromosome]['Non-syntenic, Same Chromosome'] = non_syntenic\n",
    "    transloc_chains = fetch_all_chains(chromosome, None, None, headers)\n",
    "    if chromosome == 'chr2':  # don't count chr2A and chr2B as different\n",
    "        transloc_chains = [c for c in transloc_chains if c.qName not in ['chr2A', 'chr2B']]\n",
    "    translocations = sum([e.size for chain in transloc_chains for e in chain.entries if chain.qName != chromosome])\n",
    "    chain_stats[chromosome]['Translocation from Other Chromosomes'] =  translocations\n",
    "    chain_stats[chromosome]['First Three Chains'] = sum([e.size for chain in relevant[:3] for e in chain.entries])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "df = pd.DataFrame(chain_stats) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1</th>\n",
       "      <th>chr10</th>\n",
       "      <th>chr11</th>\n",
       "      <th>chr12</th>\n",
       "      <th>chr13</th>\n",
       "      <th>chr14</th>\n",
       "      <th>chr15</th>\n",
       "      <th>chr16</th>\n",
       "      <th>chr17</th>\n",
       "      <th>chr18</th>\n",
       "      <th>...</th>\n",
       "      <th>chr22</th>\n",
       "      <th>chr3</th>\n",
       "      <th>chr4</th>\n",
       "      <th>chr5</th>\n",
       "      <th>chr6</th>\n",
       "      <th>chr7</th>\n",
       "      <th>chr8</th>\n",
       "      <th>chr9</th>\n",
       "      <th>chrX</th>\n",
       "      <th>chrY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>First Three Chains</th>\n",
       "      <td>212843994</td>\n",
       "      <td>124487884</td>\n",
       "      <td>125843504</td>\n",
       "      <td>105788978</td>\n",
       "      <td>89790006</td>\n",
       "      <td>85221470</td>\n",
       "      <td>75656180</td>\n",
       "      <td>64139961</td>\n",
       "      <td>61506966</td>\n",
       "      <td>72710459</td>\n",
       "      <td>...</td>\n",
       "      <td>31328200</td>\n",
       "      <td>191117095</td>\n",
       "      <td>149671593</td>\n",
       "      <td>95936882</td>\n",
       "      <td>164380573</td>\n",
       "      <td>144691258</td>\n",
       "      <td>126838817</td>\n",
       "      <td>107005608</td>\n",
       "      <td>109515825</td>\n",
       "      <td>9056820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Non-syntenic, Same Chromosome</th>\n",
       "      <td>104081729</td>\n",
       "      <td>39823372</td>\n",
       "      <td>78975113</td>\n",
       "      <td>63938230</td>\n",
       "      <td>31162967</td>\n",
       "      <td>513209</td>\n",
       "      <td>10205552</td>\n",
       "      <td>34476637</td>\n",
       "      <td>41881749</td>\n",
       "      <td>14511023</td>\n",
       "      <td>...</td>\n",
       "      <td>5221818</td>\n",
       "      <td>89029377</td>\n",
       "      <td>79986252</td>\n",
       "      <td>112038574</td>\n",
       "      <td>56884384</td>\n",
       "      <td>59326241</td>\n",
       "      <td>80149115</td>\n",
       "      <td>60177384</td>\n",
       "      <td>95470175</td>\n",
       "      <td>11040058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primary Chain</th>\n",
       "      <td>114984699</td>\n",
       "      <td>86709438</td>\n",
       "      <td>48114280</td>\n",
       "      <td>63230073</td>\n",
       "      <td>62370844</td>\n",
       "      <td>85021656</td>\n",
       "      <td>69081961</td>\n",
       "      <td>38655719</td>\n",
       "      <td>33475161</td>\n",
       "      <td>58288802</td>\n",
       "      <td>...</td>\n",
       "      <td>27760708</td>\n",
       "      <td>102292762</td>\n",
       "      <td>102404731</td>\n",
       "      <td>44644988</td>\n",
       "      <td>107751895</td>\n",
       "      <td>90353609</td>\n",
       "      <td>58120074</td>\n",
       "      <td>50569308</td>\n",
       "      <td>48866364</td>\n",
       "      <td>4507514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Translocation from Other Chromosomes</th>\n",
       "      <td>3834877</td>\n",
       "      <td>1315050</td>\n",
       "      <td>571079</td>\n",
       "      <td>470757</td>\n",
       "      <td>745723</td>\n",
       "      <td>924394</td>\n",
       "      <td>1200080</td>\n",
       "      <td>4884563</td>\n",
       "      <td>1417764</td>\n",
       "      <td>635990</td>\n",
       "      <td>...</td>\n",
       "      <td>2766733</td>\n",
       "      <td>625415</td>\n",
       "      <td>1545241</td>\n",
       "      <td>17266605</td>\n",
       "      <td>1045473</td>\n",
       "      <td>2340038</td>\n",
       "      <td>1021778</td>\n",
       "      <td>5328211</td>\n",
       "      <td>744372</td>\n",
       "      <td>7456790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           chr1      chr10      chr11  \\\n",
       "First Three Chains                    212843994  124487884  125843504   \n",
       "Non-syntenic, Same Chromosome         104081729   39823372   78975113   \n",
       "Primary Chain                         114984699   86709438   48114280   \n",
       "Translocation from Other Chromosomes    3834877    1315050     571079   \n",
       "\n",
       "                                          chr12     chr13     chr14     chr15  \\\n",
       "First Three Chains                    105788978  89790006  85221470  75656180   \n",
       "Non-syntenic, Same Chromosome          63938230  31162967    513209  10205552   \n",
       "Primary Chain                          63230073  62370844  85021656  69081961   \n",
       "Translocation from Other Chromosomes     470757    745723    924394   1200080   \n",
       "\n",
       "                                         chr16     chr17     chr18    ...     \\\n",
       "First Three Chains                    64139961  61506966  72710459    ...      \n",
       "Non-syntenic, Same Chromosome         34476637  41881749  14511023    ...      \n",
       "Primary Chain                         38655719  33475161  58288802    ...      \n",
       "Translocation from Other Chromosomes   4884563   1417764    635990    ...      \n",
       "\n",
       "                                         chr22       chr3       chr4  \\\n",
       "First Three Chains                    31328200  191117095  149671593   \n",
       "Non-syntenic, Same Chromosome          5221818   89029377   79986252   \n",
       "Primary Chain                         27760708  102292762  102404731   \n",
       "Translocation from Other Chromosomes   2766733     625415    1545241   \n",
       "\n",
       "                                           chr5       chr6       chr7  \\\n",
       "First Three Chains                     95936882  164380573  144691258   \n",
       "Non-syntenic, Same Chromosome         112038574   56884384   59326241   \n",
       "Primary Chain                          44644988  107751895   90353609   \n",
       "Translocation from Other Chromosomes   17266605    1045473    2340038   \n",
       "\n",
       "                                           chr8       chr9       chrX  \\\n",
       "First Three Chains                    126838817  107005608  109515825   \n",
       "Non-syntenic, Same Chromosome          80149115   60177384   95470175   \n",
       "Primary Chain                          58120074   50569308   48866364   \n",
       "Translocation from Other Chromosomes    1021778    5328211     744372   \n",
       "\n",
       "                                          chrY  \n",
       "First Three Chains                     9056820  \n",
       "Non-syntenic, Same Chromosome         11040058  \n",
       "Primary Chain                          4507514  \n",
       "Translocation from Other Chromosomes   7456790  \n",
       "\n",
       "[4 rows x 24 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "First Three Chains                      211287080\n",
       "Non-syntenic, Same Chromosome           107244972\n",
       "Primary Chain                           124837690\n",
       "Translocation from Other Chromosomes      1490781\n",
       "Name: chr2, dtype: int64"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['chr2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df.to_csv(r\"D:\\Genomes\\Human\\alignments\\hg38ToPanTro6_stats.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Non-syntenic, Same Chromosome': 112038574,\n",
       " 'Primary Chain': 44644988,\n",
       " 'Translocation from Other Chromosomes': 17266605}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chain_stats['chr5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr5 chr5 None : 811\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['chain 3886028996 chr5 181538259 + 54020836 139452659 chr5 159319378 + 43603604 117629752 23',\n",
       " 'chain 2334333576 chr5 181538259 + 18553211 46184988 chr5 159319378 - 85000998 112621309 38',\n",
       " 'chain 2207293083 chr5 181538259 + 71209725 96585715 chr5 159319378 - 115719779 141012529 41',\n",
       " 'chain 2176099837 chr5 181538259 + 155761324 181254502 chr5 159319378 + 133904970 159199629 43',\n",
       " 'chain 1568613876 chr5 181538259 + 19928 18553204 chr5 159319378 + 465 18306847 48',\n",
       " 'chain 1431413837 chr5 181538259 + 139453659 155760324 chr5 159319378 + 117630005 133905028 51',\n",
       " 'chain 32852691 chr5 181538259 + 99410290 100047704 chr5 159319378 - 81110990 82035645 145',\n",
       " 'chain 15380057 chr5 181538259 + 69789453 70364997 chr5 159319378 - 87719441 88683349 330',\n",
       " 'chain 15182291 chr5 181538259 + 70129602 70358697 chr5 159319378 + 70788969 71612852 379',\n",
       " 'chain 10279700 chr5 181538259 + 69643671 69785184 chr5 159319378 - 115715874 115850712 616']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c5 = fetch_all_chains('chr5', 'chr5', None, headers)\n",
    "[str(c) for c in c5[:10]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparison of Apes (PanTro5 and gorGor5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hg38_to_panTro4 identity 98.65%\n",
      "hg38_to_panTro4 coverage 93.26%\n",
      "\n",
      "panTro4_to_Hg38 identity 98.68%\n",
      "panTro4_to_Hg38 coverage 97.00%\n",
      "\n",
      "hg38_to_panTro5 identity 98.65%\n",
      "hg38_to_panTro5 coverage 95.44%\n",
      "\n",
      "hg38_to_gorGor4 identity 98.29%\n",
      "hg38_to_gorGor4 coverage 93.30%\n",
      "\n",
      "gorGor4_to_Hg38 identity 98.34%\n",
      "gorGor4_to_Hg38 coverage 96.15%\n",
      "\n",
      "hg38_to_gorGor5 identity 98.32%\n",
      "hg38_to_gorGor5 coverage 85.91%\n",
      "\n",
      "susie3_to_gorGor4 identity 99.71%\n",
      "susie3_to_gorGor4 coverage 96.72%\n",
      "\n"
     ]
    }
   ],
   "source": [
    "all_results = {}\n",
    "#TODO: include panTro5_to_Hg38\n",
    "for comparison in 'hg38_to_panTro4 panTro4_to_Hg38 hg38_to_panTro5 hg38_to_gorGor4 gorGor4_to_Hg38 hg38_to_gorGor5 susie3_to_gorGor4'.split():\n",
    "    all_results[comparison] = collect_stats(comparison)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'hg38_to_gorGor5_chr1': {'Aligned Variance in bp': 3014826,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9967252562573473,\n",
       "  'Alignment Coverage of Ref Chr': 0.8064999276984944,\n",
       "  'Query Gaps larger than 1000bp': 1999,\n",
       "  'Query Gaps larger than 100bp': 5310,\n",
       "  'Query Gaps larger than 10bp': 23695,\n",
       "  'Query Number of Gaps (all)': 367405,\n",
       "  'Query unique bp': 610696,\n",
       "  \"Ref Chr Total Size (No N's)\": 230472379,\n",
       "  'Ref Gaps larger than 1000bp': 1642,\n",
       "  'Ref Gaps larger than 100bp': 4977,\n",
       "  'Ref Gaps larger than 10bp': 22362,\n",
       "  'Ref Identity within Alignment': 0.9837804412756836,\n",
       "  'Ref N to query bp': 18465410,\n",
       "  'Ref Number of Gaps (all)': 219416,\n",
       "  'Ref unique bp': 44596422,\n",
       "  'Shared seq bp': 182861131,\n",
       "  'Total alignment Length': 185875957},\n",
       " 'hg38_to_gorGor5_chr10': {'Aligned Variance in bp': 1860982,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9963707325755958,\n",
       "  'Alignment Coverage of Ref Chr': 0.8360626413211497,\n",
       "  'Query Gaps larger than 1000bp': 682,\n",
       "  'Query Gaps larger than 100bp': 2544,\n",
       "  'Query Gaps larger than 10bp': 13570,\n",
       "  'Query Number of Gaps (all)': 213479,\n",
       "  'Query unique bp': 405832,\n",
       "  \"Ref Chr Total Size (No N's)\": 133262962,\n",
       "  'Ref Gaps larger than 1000bp': 801,\n",
       "  'Ref Gaps larger than 100bp': 2556,\n",
       "  'Ref Gaps larger than 10bp': 12927,\n",
       "  'Ref Identity within Alignment': 0.9832970226300337,\n",
       "  'Ref N to query bp': 524460,\n",
       "  'Ref Number of Gaps (all)': 125336,\n",
       "  'Ref unique bp': 21846778,\n",
       "  'Shared seq bp': 109555202,\n",
       "  'Total alignment Length': 111416184},\n",
       " 'hg38_to_gorGor5_chr11': {'Aligned Variance in bp': 2141386,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9967868942572597,\n",
       "  'Alignment Coverage of Ref Chr': 0.9436233274614733,\n",
       "  'Query Gaps larger than 1000bp': 695,\n",
       "  'Query Gaps larger than 100bp': 2275,\n",
       "  'Query Gaps larger than 10bp': 12622,\n",
       "  'Query Number of Gaps (all)': 209402,\n",
       "  'Query unique bp': 409216,\n",
       "  \"Ref Chr Total Size (No N's)\": 134533747,\n",
       "  'Ref Gaps larger than 1000bp': 777,\n",
       "  'Ref Gaps larger than 100bp': 2312,\n",
       "  'Ref Gaps larger than 10bp': 12273,\n",
       "  'Ref Identity within Alignment': 0.9831319432999576,\n",
       "  'Ref N to query bp': 552880,\n",
       "  'Ref Number of Gaps (all)': 120926,\n",
       "  'Ref unique bp': 7584565,\n",
       "  'Shared seq bp': 124807796,\n",
       "  'Total alignment Length': 126949182},\n",
       " 'hg38_to_gorGor5_chr12': {'Aligned Variance in bp': 2053389,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9964157002837447,\n",
       "  'Alignment Coverage of Ref Chr': 0.9554941138169838,\n",
       "  'Query Gaps larger than 1000bp': 552,\n",
       "  'Query Gaps larger than 100bp': 2267,\n",
       "  'Query Gaps larger than 10bp': 12832,\n",
       "  'Query Number of Gaps (all)': 210006,\n",
       "  'Query unique bp': 457609,\n",
       "  \"Ref Chr Total Size (No N's)\": 133138232,\n",
       "  'Ref Gaps larger than 1000bp': 673,\n",
       "  'Ref Gaps larger than 100bp': 2383,\n",
       "  'Ref Gaps larger than 10bp': 12566,\n",
       "  'Ref Identity within Alignment': 0.9838586286252318,\n",
       "  'Ref N to query bp': 137493,\n",
       "  'Ref Number of Gaps (all)': 126869,\n",
       "  'Ref unique bp': 5925435,\n",
       "  'Shared seq bp': 125159408,\n",
       "  'Total alignment Length': 127212797},\n",
       " 'hg38_to_gorGor5_chr13': {'Aligned Variance in bp': 1562216,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9972061588120467,\n",
       "  'Alignment Coverage of Ref Chr': 0.9576773857743361,\n",
       "  'Query Gaps larger than 1000bp': 468,\n",
       "  'Query Gaps larger than 100bp': 1797,\n",
       "  'Query Gaps larger than 10bp': 9279,\n",
       "  'Query Number of Gaps (all)': 153574,\n",
       "  'Query unique bp': 262898,\n",
       "  \"Ref Chr Total Size (No N's)\": 97983125,\n",
       "  'Ref Gaps larger than 1000bp': 522,\n",
       "  'Ref Gaps larger than 100bp': 1815,\n",
       "  'Ref Gaps larger than 10bp': 8962,\n",
       "  'Ref Identity within Alignment': 0.9833516743315638,\n",
       "  'Ref N to query bp': 16381203,\n",
       "  'Ref Number of Gaps (all)': 92218,\n",
       "  'Ref unique bp': 4146902,\n",
       "  'Shared seq bp': 92274007,\n",
       "  'Total alignment Length': 93836223},\n",
       " 'hg38_to_gorGor5_chr14': {'Aligned Variance in bp': 1412313,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9960795953883593,\n",
       "  'Alignment Coverage of Ref Chr': 0.9469918951326045,\n",
       "  'Query Gaps larger than 1000bp': 482,\n",
       "  'Query Gaps larger than 100bp': 1526,\n",
       "  'Query Gaps larger than 10bp': 8611,\n",
       "  'Query Number of Gaps (all)': 145460,\n",
       "  'Query unique bp': 337566,\n",
       "  \"Ref Chr Total Size (No N's)\": 90568169,\n",
       "  'Ref Gaps larger than 1000bp': 509,\n",
       "  'Ref Gaps larger than 100bp': 1573,\n",
       "  'Ref Gaps larger than 10bp': 8131,\n",
       "  'Ref Identity within Alignment': 0.9835332039398409,\n",
       "  'Ref N to query bp': 16475569,\n",
       "  'Ref Number of Gaps (all)': 85402,\n",
       "  'Ref unique bp': 4800847,\n",
       "  'Shared seq bp': 84355009,\n",
       "  'Total alignment Length': 85767322},\n",
       " 'hg38_to_gorGor5_chr15': {'Aligned Variance in bp': 1342410,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9973055829963668,\n",
       "  'Alignment Coverage of Ref Chr': 0.9197689651449972,\n",
       "  'Query Gaps larger than 1000bp': 614,\n",
       "  'Query Gaps larger than 100bp': 1707,\n",
       "  'Query Gaps larger than 10bp': 8780,\n",
       "  'Query Number of Gaps (all)': 144366,\n",
       "  'Query unique bp': 210326,\n",
       "  \"Ref Chr Total Size (No N's)\": 84640389,\n",
       "  'Ref Gaps larger than 1000bp': 653,\n",
       "  'Ref Gaps larger than 100bp': 1751,\n",
       "  'Ref Gaps larger than 10bp': 8404,\n",
       "  'Ref Identity within Alignment': 0.9827563667858397,\n",
       "  'Ref N to query bp': 17339864,\n",
       "  'Ref Number of Gaps (all)': 83588,\n",
       "  'Ref unique bp': 6790786,\n",
       "  'Shared seq bp': 76507193,\n",
       "  'Total alignment Length': 77849603},\n",
       " 'hg38_to_gorGor5_chr16': {'Aligned Variance in bp': 1445701,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9964836134020335,\n",
       "  'Alignment Coverage of Ref Chr': 0.9129876297618132,\n",
       "  'Query Gaps larger than 1000bp': 972,\n",
       "  'Query Gaps larger than 100bp': 2860,\n",
       "  'Query Gaps larger than 10bp': 11887,\n",
       "  'Query Number of Gaps (all)': 151512,\n",
       "  'Query unique bp': 263558,\n",
       "  \"Ref Chr Total Size (No N's)\": 81805943,\n",
       "  'Ref Gaps larger than 1000bp': 937,\n",
       "  'Ref Gaps larger than 100bp': 2767,\n",
       "  'Ref Gaps larger than 10bp': 11221,\n",
       "  'Ref Identity within Alignment': 0.9806434152698591,\n",
       "  'Ref N to query bp': 8422402,\n",
       "  'Ref Number of Gaps (all)': 88187,\n",
       "  'Ref unique bp': 7118129,\n",
       "  'Shared seq bp': 73242113,\n",
       "  'Total alignment Length': 74687814},\n",
       " 'hg38_to_gorGor5_chr17': {'Aligned Variance in bp': 1194221,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9971454850479905,\n",
       "  'Alignment Coverage of Ref Chr': 0.9010155806430882,\n",
       "  'Query Gaps larger than 1000bp': 519,\n",
       "  'Query Gaps larger than 100bp': 1879,\n",
       "  'Query Gaps larger than 10bp': 9571,\n",
       "  'Query Number of Gaps (all)': 149818,\n",
       "  'Query unique bp': 213876,\n",
       "  \"Ref Chr Total Size (No N's)\": 82919363,\n",
       "  'Ref Gaps larger than 1000bp': 521,\n",
       "  'Ref Gaps larger than 100bp': 1815,\n",
       "  'Ref Gaps larger than 10bp': 9319,\n",
       "  'Ref Identity within Alignment': 0.9840155960708558,\n",
       "  'Ref N to query bp': 327237,\n",
       "  'Ref Number of Gaps (all)': 87474,\n",
       "  'Ref unique bp': 8207725,\n",
       "  'Shared seq bp': 73517417,\n",
       "  'Total alignment Length': 74711638},\n",
       " 'hg38_to_gorGor5_chr18': {'Aligned Variance in bp': 904659,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9209200497833947,\n",
       "  'Alignment Coverage of Ref Chr': 0.6925950634205639,\n",
       "  'Query Gaps larger than 1000bp': 366,\n",
       "  'Query Gaps larger than 100bp': 1348,\n",
       "  'Query Gaps larger than 10bp': 7285,\n",
       "  'Query Number of Gaps (all)': 118644,\n",
       "  'Query unique bp': 4763154,\n",
       "  \"Ref Chr Total Size (No N's)\": 80088613,\n",
       "  'Ref Gaps larger than 1000bp': 386,\n",
       "  'Ref Gaps larger than 100bp': 1314,\n",
       "  'Ref Gaps larger than 10bp': 6903,\n",
       "  'Ref Identity within Alignment': 0.9836907216859124,\n",
       "  'Ref N to query bp': 173680,\n",
       "  'Ref Number of Gaps (all)': 69860,\n",
       "  'Ref unique bp': 24619635,\n",
       "  'Shared seq bp': 54564319,\n",
       "  'Total alignment Length': 55468978},\n",
       " 'hg38_to_gorGor5_chr19': {'Aligned Variance in bp': 1018842,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9845255110837265,\n",
       "  'Alignment Coverage of Ref Chr': 0.906864555042219,\n",
       "  'Query Gaps larger than 1000bp': 634,\n",
       "  'Query Gaps larger than 100bp': 2128,\n",
       "  'Query Gaps larger than 10bp': 8917,\n",
       "  'Query Number of Gaps (all)': 127907,\n",
       "  'Query unique bp': 833005,\n",
       "  \"Ref Chr Total Size (No N's)\": 58440758,\n",
       "  'Ref Gaps larger than 1000bp': 661,\n",
       "  'Ref Gaps larger than 100bp': 2066,\n",
       "  'Ref Gaps larger than 10bp': 8683,\n",
       "  'Ref Identity within Alignment': 0.9807757869130244,\n",
       "  'Ref N to query bp': 176858,\n",
       "  'Ref Number of Gaps (all)': 75732,\n",
       "  'Ref unique bp': 5442906,\n",
       "  'Shared seq bp': 51979010,\n",
       "  'Total alignment Length': 52997852},\n",
       " 'hg38_to_gorGor5_chr2': {'Aligned Variance in bp': 3528989,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9956476635814758,\n",
       "  'Alignment Coverage of Ref Chr': 0.8934407041703332,\n",
       "  'Query Gaps larger than 1000bp': 1271,\n",
       "  'Query Gaps larger than 100bp': 4303,\n",
       "  'Query Gaps larger than 10bp': 23183,\n",
       "  'Query Number of Gaps (all)': 376171,\n",
       "  'Query unique bp': 939470,\n",
       "  \"Ref Chr Total Size (No N's)\": 240547254,\n",
       "  'Ref Gaps larger than 1000bp': 1312,\n",
       "  'Ref Gaps larger than 100bp': 4142,\n",
       "  'Ref Gaps larger than 10bp': 21821,\n",
       "  'Ref Identity within Alignment': 0.983579583580664,\n",
       "  'Ref N to query bp': 1635301,\n",
       "  'Ref Number of Gaps (all)': 222391,\n",
       "  'Ref unique bp': 25632546,\n",
       "  'Shared seq bp': 211385719,\n",
       "  'Total alignment Length': 214914708},\n",
       " 'hg38_to_gorGor5_chr20': {'Aligned Variance in bp': 512218,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9835789457880038,\n",
       "  'Alignment Coverage of Ref Chr': 0.4261923851822859,\n",
       "  'Query Gaps larger than 1000bp': 344,\n",
       "  'Query Gaps larger than 100bp': 1296,\n",
       "  'Query Gaps larger than 10bp': 6829,\n",
       "  'Query Number of Gaps (all)': 106511,\n",
       "  'Query unique bp': 454987,\n",
       "  \"Ref Chr Total Size (No N's)\": 63944249,\n",
       "  'Ref Gaps larger than 1000bp': 384,\n",
       "  'Ref Gaps larger than 100bp': 1293,\n",
       "  'Ref Gaps larger than 10bp': 6560,\n",
       "  'Ref Identity within Alignment': 0.9812047693735251,\n",
       "  'Ref N to query bp': 389910,\n",
       "  'Ref Number of Gaps (all)': 59939,\n",
       "  'Ref unique bp': 36691697,\n",
       "  'Shared seq bp': 26740334,\n",
       "  'Total alignment Length': 27252552},\n",
       " 'hg38_to_gorGor5_chr21': {'Aligned Variance in bp': 729407,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9935140842662435,\n",
       "  'Alignment Coverage of Ref Chr': 0.9059751514436288,\n",
       "  'Query Gaps larger than 1000bp': 364,\n",
       "  'Query Gaps larger than 100bp': 1122,\n",
       "  'Query Gaps larger than 10bp': 5054,\n",
       "  'Query Number of Gaps (all)': 74831,\n",
       "  'Query unique bp': 237104,\n",
       "  \"Ref Chr Total Size (No N's)\": 40089009,\n",
       "  'Ref Gaps larger than 1000bp': 384,\n",
       "  'Ref Gaps larger than 100bp': 1148,\n",
       "  'Ref Gaps larger than 10bp': 5043,\n",
       "  'Ref Identity within Alignment': 0.9799170124070042,\n",
       "  'Ref N to query bp': 6621364,\n",
       "  'Ref Number of Gaps (all)': 42431,\n",
       "  'Ref unique bp': 3769363,\n",
       "  'Shared seq bp': 35590239,\n",
       "  'Total alignment Length': 36319646},\n",
       " 'hg38_to_gorGor5_chr22': {'Aligned Variance in bp': 710673,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9971813533021847,\n",
       "  'Alignment Coverage of Ref Chr': 0.9019727073088343,\n",
       "  'Query Gaps larger than 1000bp': 387,\n",
       "  'Query Gaps larger than 100bp': 1206,\n",
       "  'Query Gaps larger than 10bp': 5411,\n",
       "  'Query Number of Gaps (all)': 79219,\n",
       "  'Query unique bp': 99839,\n",
       "  \"Ref Chr Total Size (No N's)\": 39159788,\n",
       "  'Ref Gaps larger than 1000bp': 464,\n",
       "  'Ref Gaps larger than 100bp': 1355,\n",
       "  'Ref Gaps larger than 10bp': 5654,\n",
       "  'Ref Identity within Alignment': 0.9798796242241881,\n",
       "  'Ref N to query bp': 11658691,\n",
       "  'Ref Number of Gaps (all)': 44686,\n",
       "  'Ref unique bp': 3838728,\n",
       "  'Shared seq bp': 34610387,\n",
       "  'Total alignment Length': 35321060},\n",
       " 'hg38_to_gorGor5_chr3': {'Aligned Variance in bp': 3076802,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9957655575659419,\n",
       "  'Alignment Coverage of Ref Chr': 0.9635695674227467,\n",
       "  'Query Gaps larger than 1000bp': 862,\n",
       "  'Query Gaps larger than 100bp': 3001,\n",
       "  'Query Gaps larger than 10bp': 17438,\n",
       "  'Query Number of Gaps (all)': 293851,\n",
       "  'Query unique bp': 811724,\n",
       "  \"Ref Chr Total Size (No N's)\": 198100777,\n",
       "  'Ref Gaps larger than 1000bp': 897,\n",
       "  'Ref Gaps larger than 100bp': 2875,\n",
       "  'Ref Gaps larger than 10bp': 16398,\n",
       "  'Ref Identity within Alignment': 0.9838812895043835,\n",
       "  'Ref N to query bp': 195424,\n",
       "  'Ref Number of Gaps (all)': 176552,\n",
       "  'Ref unique bp': 7216897,\n",
       "  'Shared seq bp': 187807078,\n",
       "  'Total alignment Length': 190883880},\n",
       " 'hg38_to_gorGor5_chr4': {'Aligned Variance in bp': 2467657,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9964372125696941,\n",
       "  'Alignment Coverage of Ref Chr': 0.7667391522046312,\n",
       "  'Query Gaps larger than 1000bp': 1068,\n",
       "  'Query Gaps larger than 100bp': 3402,\n",
       "  'Query Gaps larger than 10bp': 17266,\n",
       "  'Query Number of Gaps (all)': 296278,\n",
       "  'Query unique bp': 520204,\n",
       "  \"Ref Chr Total Size (No N's)\": 189751874,\n",
       "  'Ref Gaps larger than 1000bp': 1046,\n",
       "  'Ref Gaps larger than 100bp': 3190,\n",
       "  'Ref Gaps larger than 10bp': 16518,\n",
       "  'Ref Identity within Alignment': 0.9830390146370761,\n",
       "  'Ref N to query bp': 451888,\n",
       "  'Ref Number of Gaps (all)': 176529,\n",
       "  'Ref unique bp': 44261683,\n",
       "  'Shared seq bp': 143022534,\n",
       "  'Total alignment Length': 145490191},\n",
       " 'hg38_to_gorGor5_chr5': {'Aligned Variance in bp': 2378154,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9960431978154152,\n",
       "  'Alignment Coverage of Ref Chr': 0.8004103050824118,\n",
       "  'Query Gaps larger than 1000bp': 902,\n",
       "  'Query Gaps larger than 100bp': 2995,\n",
       "  'Query Gaps larger than 10bp': 16514,\n",
       "  'Query Number of Gaps (all)': 271351,\n",
       "  'Query unique bp': 576359,\n",
       "  \"Ref Chr Total Size (No N's)\": 181265120,\n",
       "  'Ref Gaps larger than 1000bp': 938,\n",
       "  'Ref Gaps larger than 100bp': 2849,\n",
       "  'Ref Gaps larger than 10bp': 15395,\n",
       "  'Ref Identity within Alignment': 0.9836087127903794,\n",
       "  'Ref N to query bp': 212881,\n",
       "  'Ref Number of Gaps (all)': 163919,\n",
       "  'Ref unique bp': 36178650,\n",
       "  'Shared seq bp': 142708316,\n",
       "  'Total alignment Length': 145086470},\n",
       " 'hg38_to_gorGor5_chr6': {'Aligned Variance in bp': 2090836,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9970725454756683,\n",
       "  'Alignment Coverage of Ref Chr': 0.7689357919044005,\n",
       "  'Query Gaps larger than 1000bp': 733,\n",
       "  'Query Gaps larger than 100bp': 2680,\n",
       "  'Query Gaps larger than 10bp': 15469,\n",
       "  'Query Number of Gaps (all)': 256451,\n",
       "  'Query unique bp': 383975,\n",
       "  \"Ref Chr Total Size (No N's)\": 170078522,\n",
       "  'Ref Gaps larger than 1000bp': 801,\n",
       "  'Ref Gaps larger than 100bp': 2684,\n",
       "  'Ref Gaps larger than 10bp': 14763,\n",
       "  'Ref Identity within Alignment': 0.9840125050826979,\n",
       "  'Ref N to query bp': 667457,\n",
       "  'Ref Number of Gaps (all)': 156774,\n",
       "  'Ref unique bp': 39299059,\n",
       "  'Shared seq bp': 128688627,\n",
       "  'Total alignment Length': 130779463},\n",
       " 'hg38_to_gorGor5_chr7': {'Aligned Variance in bp': 2258999,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9967795702467271,\n",
       "  'Alignment Coverage of Ref Chr': 0.8323712183356257,\n",
       "  'Query Gaps larger than 1000bp': 1133,\n",
       "  'Query Gaps larger than 100bp': 3741,\n",
       "  'Query Gaps larger than 10bp': 17605,\n",
       "  'Query Number of Gaps (all)': 262236,\n",
       "  'Query unique bp': 427509,\n",
       "  \"Ref Chr Total Size (No N's)\": 158969389,\n",
       "  'Ref Gaps larger than 1000bp': 1307,\n",
       "  'Ref Gaps larger than 100bp': 3802,\n",
       "  'Ref Gaps larger than 10bp': 16733,\n",
       "  'Ref Identity within Alignment': 0.9829279576725616,\n",
       "  'Ref N to query bp': 365842,\n",
       "  'Ref Number of Gaps (all)': 156924,\n",
       "  'Ref unique bp': 26647845,\n",
       "  'Shared seq bp': 130062545,\n",
       "  'Total alignment Length': 132321544},\n",
       " 'hg38_to_gorGor5_chr8': {'Aligned Variance in bp': 2432640,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9958846263012341,\n",
       "  'Alignment Coverage of Ref Chr': 0.951716471547566,\n",
       "  'Query Gaps larger than 1000bp': 701,\n",
       "  'Query Gaps larger than 100bp': 2532,\n",
       "  'Query Gaps larger than 10bp': 13825,\n",
       "  'Query Number of Gaps (all)': 223359,\n",
       "  'Query unique bp': 569345,\n",
       "  \"Ref Chr Total Size (No N's)\": 144766367,\n",
       "  'Ref Gaps larger than 1000bp': 715,\n",
       "  'Ref Gaps larger than 100bp': 2477,\n",
       "  'Ref Gaps larger than 10bp': 13447,\n",
       "  'Ref Identity within Alignment': 0.9823435827999043,\n",
       "  'Ref N to query bp': 310500,\n",
       "  'Ref Number of Gaps (all)': 132335,\n",
       "  'Ref unique bp': 6989831,\n",
       "  'Shared seq bp': 135343896,\n",
       "  'Total alignment Length': 137776536},\n",
       " 'hg38_to_gorGor5_chr9': {'Aligned Variance in bp': 2021832,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9965678567478135,\n",
       "  'Alignment Coverage of Ref Chr': 0.9516192689499043,\n",
       "  'Query Gaps larger than 1000bp': 858,\n",
       "  'Query Gaps larger than 100bp': 2447,\n",
       "  'Query Gaps larger than 10bp': 12602,\n",
       "  'Query Number of Gaps (all)': 207036,\n",
       "  'Query unique bp': 399152,\n",
       "  \"Ref Chr Total Size (No N's)\": 121791380,\n",
       "  'Ref Gaps larger than 1000bp': 969,\n",
       "  'Ref Gaps larger than 100bp': 2592,\n",
       "  'Ref Gaps larger than 10bp': 12321,\n",
       "  'Ref Identity within Alignment': 0.9825552284202151,\n",
       "  'Ref N to query bp': 16604167,\n",
       "  'Ref Number of Gaps (all)': 117916,\n",
       "  'Ref unique bp': 5892356,\n",
       "  'Shared seq bp': 113877192,\n",
       "  'Total alignment Length': 115899024},\n",
       " 'hg38_to_gorGor5_chrX': {'Aligned Variance in bp': 1841583,\n",
       "  'Alignment Coverage of Query Main Chr': 0.9917573260063893,\n",
       "  'Alignment Coverage of Ref Chr': 0.7887488888753884,\n",
       "  'Query Gaps larger than 1000bp': 1197,\n",
       "  'Query Gaps larger than 100bp': 3227,\n",
       "  'Query Gaps larger than 10bp': 14050,\n",
       "  'Query Number of Gaps (all)': 206999,\n",
       "  'Query unique bp': 1015381,\n",
       "  \"Ref Chr Total Size (No N's)\": 154891493,\n",
       "  'Ref Gaps larger than 1000bp': 1252,\n",
       "  'Ref Gaps larger than 100bp': 3269,\n",
       "  'Ref Gaps larger than 10bp': 13834,\n",
       "  'Ref Identity within Alignment': 0.9849261228732211,\n",
       "  'Ref N to query bp': 1137866,\n",
       "  'Ref Number of Gaps (all)': 129243,\n",
       "  'Ref unique bp': 32721000,\n",
       "  'Shared seq bp': 120328910,\n",
       "  'Total alignment Length': 122170493}}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_results['hg38_to_gorGor5']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
