{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stats for Syntenic versus Translocation from Chain file\n",
    "We want to know how much of the alignment is a syntenic transfer 1:1 and how many bp have been rearranged between the two genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from FluentDNA.ChainFiles import chain_file_to_list, fetch_all_chains, Chain, ChainEntry\n",
    "from DNASkittleUtils.DDVUtils import pp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28612"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig_names = 'chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 chrX chrY'.split()\n",
    "chain_path = r\"D:\\Genomes\\Human\\alignments\\hg38ToPanTro6.over.chain\"\n",
    "headers = chain_file_to_list(chain_path, contig_names)  # just read the headers and discard the rest\n",
    "len(headers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'124,837,690'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pp(sum([e.size for e in headers[0].entries]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The actual coverage tends to be slightly less than the list \"query Size\" which includes gaps. It appears chain Score is not bp at all.  Likely this is a z-score."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sizes include difference already summed in.  I only know the % identity or total number of shared bp.  I don't know how that divides between my 3 categories, though if I assume it was evenly distributed, you could estimate shared bp in each category.  If I went through the unique sequence after having collapsed all ref gaps, then I'd have the exact number, since differences need not be attributed to a particular category.  We can switch to the translocations + non_syntenic + primary_chain that I've already calculated when we want to include differences.  \"No Coverage\" is already known.  \n",
    "**Conclusion**: These stats are all coverage, not shared bp or identity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2 chr2B None : 478\n",
      "chr2 chr2A None : 798\n",
      "chr1 chr1 None : 1168\n",
      "chr1 None None : 1710\n",
      "chr2 chr2 None : 0\n",
      "chr2 None None : 1610\n",
      "chr3 chr3 None : 473\n",
      "chr3 None None : 661\n",
      "chr4 chr4 None : 859\n",
      "chr4 None None : 1126\n",
      "chr5 chr5 None : 811\n",
      "chr5 None None : 1206\n",
      "chr6 chr6 None : 797\n",
      "chr6 None None : 966\n",
      "chr7 chr7 None : 1262\n",
      "chr7 None None : 1643\n",
      "chr8 chr8 None : 927\n",
      "chr8 None None : 1135\n",
      "chr9 chr9 None : 588\n",
      "chr9 None None : 996\n",
      "chr10 chr10 None : 929\n",
      "chr10 None None : 1209\n",
      "chr11 chr11 None : 554\n",
      "chr11 None None : 701\n",
      "chr12 chr12 None : 613\n",
      "chr12 None None : 719\n",
      "chr13 chr13 None : 529\n",
      "chr13 None None : 701\n",
      "chr14 chr14 None : 477\n",
      "chr14 None None : 631\n",
      "chr15 chr15 None : 263\n",
      "chr15 None None : 461\n",
      "chr16 chr16 None : 752\n",
      "chr16 None None : 1105\n",
      "chr17 chr17 None : 748\n",
      "chr17 None None : 1003\n",
      "chr18 chr18 None : 586\n",
      "chr18 None None : 677\n",
      "chr19 chr19 None : 818\n",
      "chr19 None None : 986\n",
      "chr20 chr20 None : 505\n",
      "chr20 None None : 616\n",
      "chr21 chr21 None : 298\n",
      "chr21 None None : 541\n",
      "chr22 chr22 None : 530\n",
      "chr22 None None : 777\n",
      "chrX chrX None : 590\n",
      "chrX None None : 1488\n",
      "chrY chrY None : 313\n",
      "chrY None None : 1592\n"
     ]
    }
   ],
   "source": [
    "chain_stats = {name: {} for name in contig_names}\n",
    "chr2_chains = fetch_all_chains('chr2', 'chr2B', None, headers) +\\\n",
    "            fetch_all_chains('chr2', 'chr2A', None, headers)  # special case chr2\n",
    "chr2_chains.sort(key=lambda chain: -chain.score)\n",
    "\n",
    "for chromosome in contig_names:\n",
    "    relevant = fetch_all_chains(chromosome, chromosome, None, headers)\n",
    "    if chromosome == 'chr2':\n",
    "        relevant = chr2_chains\n",
    "    chain_stats[chromosome]['Primary Chain'] = sum([e.size for e in relevant[0].entries])\n",
    "    non_syntenic = sum([e.size for chain in relevant[1:] for e in chain.entries]) if len(relevant) > 1 else 0\n",
    "    chain_stats[chromosome]['Non-syntenic Same Chromosome'] = non_syntenic\n",
    "    transloc_chains = fetch_all_chains(chromosome, None, None, headers)\n",
    "    if chromosome == 'chr2':  # don't count chr2A and chr2B as different\n",
    "        transloc_chains = [c for c in transloc_chains if c.qName not in ['chr2A', 'chr2B']]\n",
    "    translocations = sum([e.size for chain in transloc_chains for e in chain.entries if chain.qName != chromosome])\n",
    "    chain_stats[chromosome]['Translocation from Other Chromosomes'] =  translocations\n",
    "    chain_stats[chromosome]['First Two Chains'] = sum([e.size for chain in relevant[:2] for e in chain.entries])\n",
    "    chain_stats[chromosome]['First Three Chains'] = sum([e.size for chain in relevant[:3] for e in chain.entries])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1</th>\n",
       "      <th>chr10</th>\n",
       "      <th>chr11</th>\n",
       "      <th>chr12</th>\n",
       "      <th>chr13</th>\n",
       "      <th>chr14</th>\n",
       "      <th>chr15</th>\n",
       "      <th>chr16</th>\n",
       "      <th>chr17</th>\n",
       "      <th>chr18</th>\n",
       "      <th>...</th>\n",
       "      <th>chr22</th>\n",
       "      <th>chr3</th>\n",
       "      <th>chr4</th>\n",
       "      <th>chr5</th>\n",
       "      <th>chr6</th>\n",
       "      <th>chr7</th>\n",
       "      <th>chr8</th>\n",
       "      <th>chr9</th>\n",
       "      <th>chrX</th>\n",
       "      <th>chrY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>First Three Chains</th>\n",
       "      <td>212843994</td>\n",
       "      <td>124487884</td>\n",
       "      <td>125843504</td>\n",
       "      <td>105788978</td>\n",
       "      <td>89790006</td>\n",
       "      <td>85221470</td>\n",
       "      <td>75656180</td>\n",
       "      <td>64139961</td>\n",
       "      <td>61506966</td>\n",
       "      <td>72710459</td>\n",
       "      <td>...</td>\n",
       "      <td>31328200</td>\n",
       "      <td>191117095</td>\n",
       "      <td>149671593</td>\n",
       "      <td>95936882</td>\n",
       "      <td>164380573</td>\n",
       "      <td>144691258</td>\n",
       "      <td>126838817</td>\n",
       "      <td>107005608</td>\n",
       "      <td>109515825</td>\n",
       "      <td>9056820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>First Two Chains</th>\n",
       "      <td>188398140</td>\n",
       "      <td>123980669</td>\n",
       "      <td>94011435</td>\n",
       "      <td>92381491</td>\n",
       "      <td>87445742</td>\n",
       "      <td>85180600</td>\n",
       "      <td>73889080</td>\n",
       "      <td>55223212</td>\n",
       "      <td>51635953</td>\n",
       "      <td>72705936</td>\n",
       "      <td>...</td>\n",
       "      <td>29855569</td>\n",
       "      <td>191036976</td>\n",
       "      <td>127701947</td>\n",
       "      <td>71133305</td>\n",
       "      <td>161361608</td>\n",
       "      <td>140387084</td>\n",
       "      <td>96497462</td>\n",
       "      <td>89454020</td>\n",
       "      <td>82867553</td>\n",
       "      <td>7299314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Non-syntenic Same Chromosome</th>\n",
       "      <td>104081729</td>\n",
       "      <td>39823372</td>\n",
       "      <td>78975113</td>\n",
       "      <td>63938230</td>\n",
       "      <td>31162967</td>\n",
       "      <td>513209</td>\n",
       "      <td>10205552</td>\n",
       "      <td>34476637</td>\n",
       "      <td>41881749</td>\n",
       "      <td>14511023</td>\n",
       "      <td>...</td>\n",
       "      <td>5221818</td>\n",
       "      <td>89029377</td>\n",
       "      <td>79986252</td>\n",
       "      <td>112038574</td>\n",
       "      <td>56884384</td>\n",
       "      <td>59326241</td>\n",
       "      <td>80149115</td>\n",
       "      <td>60177384</td>\n",
       "      <td>95470175</td>\n",
       "      <td>11040058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primary Chain</th>\n",
       "      <td>114984699</td>\n",
       "      <td>86709438</td>\n",
       "      <td>48114280</td>\n",
       "      <td>63230073</td>\n",
       "      <td>62370844</td>\n",
       "      <td>85021656</td>\n",
       "      <td>69081961</td>\n",
       "      <td>38655719</td>\n",
       "      <td>33475161</td>\n",
       "      <td>58288802</td>\n",
       "      <td>...</td>\n",
       "      <td>27760708</td>\n",
       "      <td>102292762</td>\n",
       "      <td>102404731</td>\n",
       "      <td>44644988</td>\n",
       "      <td>107751895</td>\n",
       "      <td>90353609</td>\n",
       "      <td>58120074</td>\n",
       "      <td>50569308</td>\n",
       "      <td>48866364</td>\n",
       "      <td>4507514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Translocation from Other Chromosomes</th>\n",
       "      <td>3834877</td>\n",
       "      <td>1315050</td>\n",
       "      <td>571079</td>\n",
       "      <td>470757</td>\n",
       "      <td>745723</td>\n",
       "      <td>924394</td>\n",
       "      <td>1200080</td>\n",
       "      <td>4884563</td>\n",
       "      <td>1417764</td>\n",
       "      <td>635990</td>\n",
       "      <td>...</td>\n",
       "      <td>2766733</td>\n",
       "      <td>625415</td>\n",
       "      <td>1545241</td>\n",
       "      <td>17266605</td>\n",
       "      <td>1045473</td>\n",
       "      <td>2340038</td>\n",
       "      <td>1021778</td>\n",
       "      <td>5328211</td>\n",
       "      <td>744372</td>\n",
       "      <td>7456790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           chr1      chr10      chr11  \\\n",
       "First Three Chains                    212843994  124487884  125843504   \n",
       "First Two Chains                      188398140  123980669   94011435   \n",
       "Non-syntenic Same Chromosome          104081729   39823372   78975113   \n",
       "Primary Chain                         114984699   86709438   48114280   \n",
       "Translocation from Other Chromosomes    3834877    1315050     571079   \n",
       "\n",
       "                                          chr12     chr13     chr14     chr15  \\\n",
       "First Three Chains                    105788978  89790006  85221470  75656180   \n",
       "First Two Chains                       92381491  87445742  85180600  73889080   \n",
       "Non-syntenic Same Chromosome           63938230  31162967    513209  10205552   \n",
       "Primary Chain                          63230073  62370844  85021656  69081961   \n",
       "Translocation from Other Chromosomes     470757    745723    924394   1200080   \n",
       "\n",
       "                                         chr16     chr17     chr18    ...     \\\n",
       "First Three Chains                    64139961  61506966  72710459    ...      \n",
       "First Two Chains                      55223212  51635953  72705936    ...      \n",
       "Non-syntenic Same Chromosome          34476637  41881749  14511023    ...      \n",
       "Primary Chain                         38655719  33475161  58288802    ...      \n",
       "Translocation from Other Chromosomes   4884563   1417764    635990    ...      \n",
       "\n",
       "                                         chr22       chr3       chr4  \\\n",
       "First Three Chains                    31328200  191117095  149671593   \n",
       "First Two Chains                      29855569  191036976  127701947   \n",
       "Non-syntenic Same Chromosome           5221818   89029377   79986252   \n",
       "Primary Chain                         27760708  102292762  102404731   \n",
       "Translocation from Other Chromosomes   2766733     625415    1545241   \n",
       "\n",
       "                                           chr5       chr6       chr7  \\\n",
       "First Three Chains                     95936882  164380573  144691258   \n",
       "First Two Chains                       71133305  161361608  140387084   \n",
       "Non-syntenic Same Chromosome          112038574   56884384   59326241   \n",
       "Primary Chain                          44644988  107751895   90353609   \n",
       "Translocation from Other Chromosomes   17266605    1045473    2340038   \n",
       "\n",
       "                                           chr8       chr9       chrX  \\\n",
       "First Three Chains                    126838817  107005608  109515825   \n",
       "First Two Chains                       96497462   89454020   82867553   \n",
       "Non-syntenic Same Chromosome           80149115   60177384   95470175   \n",
       "Primary Chain                          58120074   50569308   48866364   \n",
       "Translocation from Other Chromosomes    1021778    5328211     744372   \n",
       "\n",
       "                                          chrY  \n",
       "First Three Chains                     9056820  \n",
       "First Two Chains                       7299314  \n",
       "Non-syntenic Same Chromosome          11040058  \n",
       "Primary Chain                          4507514  \n",
       "Translocation from Other Chromosomes   7456790  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd \n",
    "df = pd.DataFrame(chain_stats) \n",
    "df.to_csv(r\"D:\\Genomes\\Human\\alignments\\hg38ToPanTro6_stats.csv\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Double Check sequence lengths with Ns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from DNASkittleUtils.Contigs import Contig, read_contigs\n",
    "hg38 = read_contigs(r\"D:\\Genomes\\Human\\hg38.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1</th>\n",
       "      <th>chr2</th>\n",
       "      <th>chr3</th>\n",
       "      <th>chr4</th>\n",
       "      <th>chr5</th>\n",
       "      <th>chr6</th>\n",
       "      <th>chr7</th>\n",
       "      <th>chr8</th>\n",
       "      <th>chr9</th>\n",
       "      <th>chr10</th>\n",
       "      <th>...</th>\n",
       "      <th>chr15</th>\n",
       "      <th>chr16</th>\n",
       "      <th>chr17</th>\n",
       "      <th>chr18</th>\n",
       "      <th>chr19</th>\n",
       "      <th>chr20</th>\n",
       "      <th>chr21</th>\n",
       "      <th>chr22</th>\n",
       "      <th>chrX</th>\n",
       "      <th>chrY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>First Three Chains</th>\n",
       "      <td>212843994</td>\n",
       "      <td>211287080</td>\n",
       "      <td>191117095</td>\n",
       "      <td>149671593</td>\n",
       "      <td>95936882</td>\n",
       "      <td>164380573</td>\n",
       "      <td>144691258</td>\n",
       "      <td>126838817</td>\n",
       "      <td>107005608</td>\n",
       "      <td>124487884</td>\n",
       "      <td>...</td>\n",
       "      <td>75656180</td>\n",
       "      <td>64139961</td>\n",
       "      <td>61506966</td>\n",
       "      <td>72710459</td>\n",
       "      <td>51430338</td>\n",
       "      <td>58095027</td>\n",
       "      <td>32627633</td>\n",
       "      <td>31328200</td>\n",
       "      <td>109515825</td>\n",
       "      <td>9056820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>First Two Chains</th>\n",
       "      <td>188398140</td>\n",
       "      <td>195592657</td>\n",
       "      <td>191036976</td>\n",
       "      <td>127701947</td>\n",
       "      <td>71133305</td>\n",
       "      <td>161361608</td>\n",
       "      <td>140387084</td>\n",
       "      <td>96497462</td>\n",
       "      <td>89454020</td>\n",
       "      <td>123980669</td>\n",
       "      <td>...</td>\n",
       "      <td>73889080</td>\n",
       "      <td>55223212</td>\n",
       "      <td>51635953</td>\n",
       "      <td>72705936</td>\n",
       "      <td>50516523</td>\n",
       "      <td>57763891</td>\n",
       "      <td>32473339</td>\n",
       "      <td>29855569</td>\n",
       "      <td>82867553</td>\n",
       "      <td>7299314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Non-syntenic Same Chromosome</th>\n",
       "      <td>104081729</td>\n",
       "      <td>107244972</td>\n",
       "      <td>89029377</td>\n",
       "      <td>79986252</td>\n",
       "      <td>112038574</td>\n",
       "      <td>56884384</td>\n",
       "      <td>59326241</td>\n",
       "      <td>80149115</td>\n",
       "      <td>60177384</td>\n",
       "      <td>39823372</td>\n",
       "      <td>...</td>\n",
       "      <td>10205552</td>\n",
       "      <td>34476637</td>\n",
       "      <td>41881749</td>\n",
       "      <td>14511023</td>\n",
       "      <td>24405460</td>\n",
       "      <td>26698533</td>\n",
       "      <td>783172</td>\n",
       "      <td>5221818</td>\n",
       "      <td>95470175</td>\n",
       "      <td>11040058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primary Chain</th>\n",
       "      <td>114984699</td>\n",
       "      <td>124837690</td>\n",
       "      <td>102292762</td>\n",
       "      <td>102404731</td>\n",
       "      <td>44644988</td>\n",
       "      <td>107751895</td>\n",
       "      <td>90353609</td>\n",
       "      <td>58120074</td>\n",
       "      <td>50569308</td>\n",
       "      <td>86709438</td>\n",
       "      <td>...</td>\n",
       "      <td>69081961</td>\n",
       "      <td>38655719</td>\n",
       "      <td>33475161</td>\n",
       "      <td>58288802</td>\n",
       "      <td>28339269</td>\n",
       "      <td>32528719</td>\n",
       "      <td>32285226</td>\n",
       "      <td>27760708</td>\n",
       "      <td>48866364</td>\n",
       "      <td>4507514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total Length</th>\n",
       "      <td>248956422</td>\n",
       "      <td>242193529</td>\n",
       "      <td>198295559</td>\n",
       "      <td>190214555</td>\n",
       "      <td>181538259</td>\n",
       "      <td>170805979</td>\n",
       "      <td>159345973</td>\n",
       "      <td>145138636</td>\n",
       "      <td>138394717</td>\n",
       "      <td>133797422</td>\n",
       "      <td>...</td>\n",
       "      <td>101991189</td>\n",
       "      <td>90338345</td>\n",
       "      <td>83257441</td>\n",
       "      <td>80373285</td>\n",
       "      <td>58617616</td>\n",
       "      <td>64444167</td>\n",
       "      <td>46709983</td>\n",
       "      <td>50818468</td>\n",
       "      <td>156040895</td>\n",
       "      <td>57227415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total Length (No Ns)</th>\n",
       "      <td>230481012</td>\n",
       "      <td>240548228</td>\n",
       "      <td>198100135</td>\n",
       "      <td>189752667</td>\n",
       "      <td>181265378</td>\n",
       "      <td>170078522</td>\n",
       "      <td>158970131</td>\n",
       "      <td>144768136</td>\n",
       "      <td>121790550</td>\n",
       "      <td>133262962</td>\n",
       "      <td>...</td>\n",
       "      <td>84641325</td>\n",
       "      <td>81805943</td>\n",
       "      <td>82920204</td>\n",
       "      <td>80089605</td>\n",
       "      <td>58440758</td>\n",
       "      <td>63944257</td>\n",
       "      <td>40088619</td>\n",
       "      <td>39159777</td>\n",
       "      <td>154893029</td>\n",
       "      <td>26415043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Translocation from Other Chromosomes</th>\n",
       "      <td>3834877</td>\n",
       "      <td>1490781</td>\n",
       "      <td>625415</td>\n",
       "      <td>1545241</td>\n",
       "      <td>17266605</td>\n",
       "      <td>1045473</td>\n",
       "      <td>2340038</td>\n",
       "      <td>1021778</td>\n",
       "      <td>5328211</td>\n",
       "      <td>1315050</td>\n",
       "      <td>...</td>\n",
       "      <td>1200080</td>\n",
       "      <td>4884563</td>\n",
       "      <td>1417764</td>\n",
       "      <td>635990</td>\n",
       "      <td>444370</td>\n",
       "      <td>395480</td>\n",
       "      <td>3390981</td>\n",
       "      <td>2766733</td>\n",
       "      <td>744372</td>\n",
       "      <td>7456790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           chr1       chr2       chr3  \\\n",
       "First Three Chains                    212843994  211287080  191117095   \n",
       "First Two Chains                      188398140  195592657  191036976   \n",
       "Non-syntenic Same Chromosome          104081729  107244972   89029377   \n",
       "Primary Chain                         114984699  124837690  102292762   \n",
       "Total Length                          248956422  242193529  198295559   \n",
       "Total Length (No Ns)                  230481012  240548228  198100135   \n",
       "Translocation from Other Chromosomes    3834877    1490781     625415   \n",
       "\n",
       "                                           chr4       chr5       chr6  \\\n",
       "First Three Chains                    149671593   95936882  164380573   \n",
       "First Two Chains                      127701947   71133305  161361608   \n",
       "Non-syntenic Same Chromosome           79986252  112038574   56884384   \n",
       "Primary Chain                         102404731   44644988  107751895   \n",
       "Total Length                          190214555  181538259  170805979   \n",
       "Total Length (No Ns)                  189752667  181265378  170078522   \n",
       "Translocation from Other Chromosomes    1545241   17266605    1045473   \n",
       "\n",
       "                                           chr7       chr8       chr9  \\\n",
       "First Three Chains                    144691258  126838817  107005608   \n",
       "First Two Chains                      140387084   96497462   89454020   \n",
       "Non-syntenic Same Chromosome           59326241   80149115   60177384   \n",
       "Primary Chain                          90353609   58120074   50569308   \n",
       "Total Length                          159345973  145138636  138394717   \n",
       "Total Length (No Ns)                  158970131  144768136  121790550   \n",
       "Translocation from Other Chromosomes    2340038    1021778    5328211   \n",
       "\n",
       "                                          chr10    ...         chr15  \\\n",
       "First Three Chains                    124487884    ...      75656180   \n",
       "First Two Chains                      123980669    ...      73889080   \n",
       "Non-syntenic Same Chromosome           39823372    ...      10205552   \n",
       "Primary Chain                          86709438    ...      69081961   \n",
       "Total Length                          133797422    ...     101991189   \n",
       "Total Length (No Ns)                  133262962    ...      84641325   \n",
       "Translocation from Other Chromosomes    1315050    ...       1200080   \n",
       "\n",
       "                                         chr16     chr17     chr18     chr19  \\\n",
       "First Three Chains                    64139961  61506966  72710459  51430338   \n",
       "First Two Chains                      55223212  51635953  72705936  50516523   \n",
       "Non-syntenic Same Chromosome          34476637  41881749  14511023  24405460   \n",
       "Primary Chain                         38655719  33475161  58288802  28339269   \n",
       "Total Length                          90338345  83257441  80373285  58617616   \n",
       "Total Length (No Ns)                  81805943  82920204  80089605  58440758   \n",
       "Translocation from Other Chromosomes   4884563   1417764    635990    444370   \n",
       "\n",
       "                                         chr20     chr21     chr22       chrX  \\\n",
       "First Three Chains                    58095027  32627633  31328200  109515825   \n",
       "First Two Chains                      57763891  32473339  29855569   82867553   \n",
       "Non-syntenic Same Chromosome          26698533    783172   5221818   95470175   \n",
       "Primary Chain                         32528719  32285226  27760708   48866364   \n",
       "Total Length                          64444167  46709983  50818468  156040895   \n",
       "Total Length (No Ns)                  63944257  40088619  39159777  154893029   \n",
       "Translocation from Other Chromosomes    395480   3390981   2766733     744372   \n",
       "\n",
       "                                          chrY  \n",
       "First Three Chains                     9056820  \n",
       "First Two Chains                       7299314  \n",
       "Non-syntenic Same Chromosome          11040058  \n",
       "Primary Chain                          4507514  \n",
       "Total Length                          57227415  \n",
       "Total Length (No Ns)                  26415043  \n",
       "Translocation from Other Chromosomes   7456790  \n",
       "\n",
       "[7 rows x 24 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for contig in hg38:\n",
    "    if contig.name in contig_names:\n",
    "        chain_stats[contig.name]['Total Length'] = len(contig.seq)\n",
    "        chain_stats[contig.name]['Total Length (No Ns)'] = len(contig.seq) - contig.seq.count('N') - contig.seq.count('n')\n",
    "\n",
    "df = pd.DataFrame(chain_stats) \n",
    "df = df[contig_names] # reorder columns\n",
    "df.to_csv(r\"D:\\Genomes\\Human\\alignments\\hg38ToPanTro6_stats.csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gather Alignment Stats into a single File\n",
    "This functionality has been partially integrated into FluentDNA for derived stats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Specify your target directory\n",
    "# Traverse directory\n",
    "# parse files to dict structure\n",
    "# Compute derivative values\n",
    "# Print medians"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "from os.path import dirname, join\n",
    "from glob import glob \n",
    "from statistics import median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def print_median(name, label, key, contents):\n",
    "    print(name, label, \"{:0.2%}\".format(median([contents[d][key] for d in contents.keys()])))\n",
    "    \n",
    "def print_average(name, label, key, contents):\n",
    "    avg = 0\n",
    "    total_length = 0\n",
    "    genome_length = sum([v[\"Ref Chr Total Size (No N's)\"] for k,v in contents.items()])\n",
    "    for chr in contents.keys():\n",
    "        avg += contents[chr][key] * contents[chr][\"Ref Chr Total Size (No N's)\"] / genome_length\n",
    "    print(name, label, \"{:0.2%}\".format(avg))\n",
    "\n",
    "def summarize_stats(name, contents):\n",
    "#     print(\"Median Chromosome Value\")\n",
    "#     print_median(name, 'identity', 'Ref Identity within Alignment', contents)\n",
    "#     print_median(name, 'coverage', 'Alignment Coverage of Ref Chr', contents)\n",
    "#     print_median(name, 'query coverage', \"Alignment Coverage of Query Main Chr\", contents)\n",
    "    \n",
    "    print_average(name, 'identity', 'Ref Identity within Alignment', contents)\n",
    "    print_average(name, 'coverage', 'Alignment Coverage of Ref Chr', contents)\n",
    "#     print_average(name, 'query coverage', \"Alignment Coverage of Query Main Chr\", contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def number(string):\n",
    "    try:\n",
    "        return int(string)\n",
    "    except ValueError:\n",
    "        return float(string)\n",
    "    \n",
    "def collect_stats(folder, prefix=''):\n",
    "    contents = {}\n",
    "      # Specify you target directory\n",
    "    os.chdir(folder)\n",
    "    # Traverse directory\n",
    "    subdirs = glob(prefix + '*')\n",
    "    for d in subdirs:\n",
    "        stats_file = os.path.join(d, r'stats.txt')\n",
    "        if os.path.exists(stats_file):\n",
    "            with open(stats_file) as stat:\n",
    "                # parse files to dict structure\n",
    "                lines = [line.split('\\t') for line in stat.read().splitlines()]\n",
    "                v = {line[0]: number(line[1]) for line in lines if len(line) == 2}\n",
    "                # Compute derivative values\n",
    "                v[\"Total alignment Length\"] = v['Aligned Variance in bp'] + v['Shared seq bp']\n",
    "                v[\"Ref Chr Total Size (No N's)\"] = v['Ref unique bp'] + v[\"Total alignment Length\"]\n",
    "                v[\"Ref Identity within Alignment\"] = 1 - (v['Aligned Variance in bp'] / v['Total alignment Length'])\n",
    "                v[\"Alignment Coverage of Ref Chr\"] = v[\"Total alignment Length\"] / v[\"Ref Chr Total Size (No N's)\"]\n",
    "                v[\"Alignment Coverage of Query Main Chr\"] = v['Total alignment Length'] / (v['Total alignment Length'] + v[\"Query unique bp\"])\n",
    "                contents[d] = v\n",
    "    # Print medians\n",
    "    name = prefix or os.path.basename(folder)\n",
    "#     print(\"Averages Weighted by Chromosome Length\")\n",
    "    summarize_stats(name, contents)\n",
    "    return contents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Human Hg38 vs Chimpanzee PanTro6_ identity 98.65%\n",
      "Human Hg38 vs Chimpanzee PanTro6_ coverage 94.95%\n",
      "\n",
      "\n",
      "chr1\n",
      "Ref Number of Gaps (all)\t165609\n",
      "Ref Gaps larger than 10bp\t18763\n",
      "Query Number of Gaps (all)\t171053\n",
      "Query Gaps larger than 10bp\t21270\n",
      "Ref Gaps larger than 100bp\t4479\n",
      "Query Gaps larger than 100bp\t4832\n",
      "Ref Gaps larger than 1000bp\t1655\n",
      "Query Gaps larger than 1000bp\t1943\n",
      "Ref N to query bp\t18475471\n",
      "Ref unique bp\t8720120\n",
      "Shared seq bp\t220917707\n",
      "Aligned Variance in bp\t2830836\n",
      "Query unique bp\t8724657\n",
      "Query N to ref in bp\t2495044\n",
      "Total alignment Length\t223748543\n",
      "Ref Chr Total Size (No N's)\t232468663\n",
      "Ref Identity within Alignment\t0.9873481366088717\n",
      "Alignment Coverage of Ref Chr\t0.9624890516964001\n",
      "Alignment Coverage of Query Main Chr\t0.9624702675405165\n",
      "\n",
      "\n",
      "chr10\n",
      "Query Number of Gaps (all)\t98327\n",
      "Query Gaps larger than 10bp\t12245\n",
      "Ref Number of Gaps (all)\t95484\n",
      "Ref Gaps larger than 10bp\t11050\n",
      "Query Gaps larger than 100bp\t2354\n",
      "Query Gaps larger than 1000bp\t620\n",
      "Ref Gaps larger than 100bp\t2504\n",
      "Ref Gaps larger than 1000bp\t819\n",
      "Ref N to query bp\t534460\n",
      "Ref unique bp\t6082908\n",
      "Shared seq bp\t126144862\n",
      "Aligned Variance in bp\t1708138\n",
      "Query unique bp\t4145888\n",
      "Query N to ref in bp\t1197562\n",
      "Total alignment Length\t127853000\n",
      "Ref Chr Total Size (No N's)\t133935908\n",
      "Ref Identity within Alignment\t0.986639828553104\n",
      "Alignment Coverage of Ref Chr\t0.9545834415069632\n",
      "Alignment Coverage of Query Main Chr\t0.9685914929828803\n",
      "\n",
      "\n",
      "chr11\n",
      "Query Number of Gaps (all)\t94230\n",
      "Ref Number of Gaps (all)\t92132\n",
      "Query Gaps larger than 10bp\t10997\n",
      "Query Gaps larger than 100bp\t1996\n",
      "Query Gaps larger than 1000bp\t598\n",
      "Ref Gaps larger than 10bp\t9946\n",
      "Ref Gaps larger than 100bp\t1971\n",
      "Ref Gaps larger than 1000bp\t677\n",
      "Ref N to query bp\t552880\n",
      "Shared seq bp\t126357568\n",
      "Aligned Variance in bp\t1694144\n",
      "Ref unique bp\t7284509\n",
      "Query unique bp\t2765835\n",
      "Query N to ref in bp\t8258\n",
      "Total alignment Length\t128051712\n",
      "Ref Chr Total Size (No N's)\t135336221\n",
      "Ref Identity within Alignment\t0.986769844982627\n",
      "Alignment Coverage of Ref Chr\t0.9461747273111756\n",
      "Alignment Coverage of Query Main Chr\t0.9788573087981844\n",
      "\n",
      "\n",
      "chr12\n",
      "Ref Number of Gaps (all)\t96419\n",
      "Query Number of Gaps (all)\t98479\n",
      "Ref Gaps larger than 10bp\t9907\n",
      "Ref Gaps larger than 100bp\t1901\n",
      "Query Gaps larger than 10bp\t11165\n",
      "Ref Gaps larger than 1000bp\t586\n",
      "Query Gaps larger than 100bp\t1887\n",
      "Query Gaps larger than 1000bp\t465\n",
      "Ref N to query bp\t137493\n",
      "Ref unique bp\t6103971\n",
      "Shared seq bp\t126059664\n",
      "Aligned Variance in bp\t1618815\n",
      "Query unique bp\t1956826\n",
      "Query N to ref in bp\t5774\n",
      "Total alignment Length\t127678479\n",
      "Ref Chr Total Size (No N's)\t133782450\n",
      "Ref Identity within Alignment\t0.987321160052353\n",
      "Alignment Coverage of Ref Chr\t0.9543739033034602\n",
      "Alignment Coverage of Query Main Chr\t0.9849051460171286\n",
      "\n",
      "\n",
      "chr13\n",
      "Ref Number of Gaps (all)\t71882\n",
      "Query Number of Gaps (all)\t74740\n",
      "Query Gaps larger than 10bp\t8273\n",
      "Query Gaps larger than 100bp\t1586\n",
      "Ref Gaps larger than 10bp\t7416\n",
      "Ref Gaps larger than 100bp\t1638\n",
      "Ref Gaps larger than 1000bp\t506\n",
      "Query Gaps larger than 1000bp\t408\n",
      "Ref N to query bp\t16381203\n",
      "Ref unique bp\t3849601\n",
      "Shared seq bp\t93347690\n",
      "Aligned Variance in bp\t1250938\n",
      "Query unique bp\t5205275\n",
      "Query N to ref in bp\t402562\n",
      "Total alignment Length\t94598628\n",
      "Ref Chr Total Size (No N's)\t98448229\n",
      "Ref Identity within Alignment\t0.9867763621265205\n",
      "Alignment Coverage of Ref Chr\t0.9608972041538706\n",
      "Alignment Coverage of Query Main Chr\t0.9478449755617273\n",
      "\n",
      "\n",
      "chr14\n",
      "Query Number of Gaps (all)\t66865\n",
      "Query Gaps larger than 10bp\t7557\n",
      "Ref Number of Gaps (all)\t65255\n",
      "Query Gaps larger than 100bp\t1328\n",
      "Query Gaps larger than 1000bp\t440\n",
      "Ref Gaps larger than 10bp\t6801\n",
      "Ref Gaps larger than 100bp\t1370\n",
      "Ref Gaps larger than 1000bp\t462\n",
      "Ref N to query bp\t16475569\n",
      "Ref unique bp\t4245909\n",
      "Shared seq bp\t85325045\n",
      "Query unique bp\t2098847\n",
      "Aligned Variance in bp\t1134250\n",
      "Query N to ref in bp\t475733\n",
      "Total alignment Length\t86459295\n",
      "Ref Chr Total Size (No N's)\t90705204\n",
      "Ref Identity within Alignment\t0.9868811097754151\n",
      "Alignment Coverage of Ref Chr\t0.9531900176311825\n",
      "Alignment Coverage of Query Main Chr\t0.9762997850609829\n",
      "\n",
      "\n",
      "chr15\n",
      "Query Number of Gaps (all)\t66261\n",
      "Ref Number of Gaps (all)\t63744\n",
      "Query Gaps larger than 10bp\t7551\n",
      "Ref Gaps larger than 10bp\t6843\n",
      "Query Gaps larger than 100bp\t1404\n",
      "Query Gaps larger than 1000bp\t478\n",
      "Ref Gaps larger than 100bp\t1455\n",
      "Ref Gaps larger than 1000bp\t578\n",
      "Ref N to query bp\t17349864\n",
      "Ref unique bp\t4447223\n",
      "Shared seq bp\t79377739\n",
      "Aligned Variance in bp\t1110620\n",
      "Query unique bp\t9776198\n",
      "Query N to ref in bp\t1518735\n",
      "Total alignment Length\t80488359\n",
      "Ref Chr Total Size (No N's)\t84935582\n",
      "Ref Identity within Alignment\t0.9862014828753062\n",
      "Alignment Coverage of Ref Chr\t0.9476400479601117\n",
      "Alignment Coverage of Query Main Chr\t0.8916939458308093\n",
      "\n",
      "\n",
      "chr16\n",
      "Query Number of Gaps (all)\t70208\n",
      "Query Gaps larger than 10bp\t10578\n",
      "Ref Number of Gaps (all)\t66458\n",
      "Ref Gaps larger than 10bp\t9996\n",
      "Ref Gaps larger than 100bp\t2892\n",
      "Query Gaps larger than 100bp\t2723\n",
      "Ref Gaps larger than 1000bp\t1127\n",
      "Query Gaps larger than 1000bp\t969\n",
      "Ref N to query bp\t8533476\n",
      "Ref unique bp\t4824479\n",
      "Shared seq bp\t76802346\n",
      "Query unique bp\t8134807\n",
      "Aligned Variance in bp\t1219533\n",
      "Query N to ref in bp\t355075\n",
      "Total alignment Length\t78021879\n",
      "Ref Chr Total Size (No N's)\t82846358\n",
      "Ref Identity within Alignment\t0.9843693459369263\n",
      "Alignment Coverage of Ref Chr\t0.9417659494458405\n",
      "Alignment Coverage of Query Main Chr\t0.9055812453139156\n",
      "\n",
      "\n",
      "chr17\n",
      "Query Number of Gaps (all)\t64542\n",
      "Ref Number of Gaps (all)\t64312\n",
      "Query Gaps larger than 10bp\t8707\n",
      "Ref Gaps larger than 10bp\t8164\n",
      "Query Gaps larger than 100bp\t1727\n",
      "Ref Gaps larger than 100bp\t1826\n",
      "Ref Gaps larger than 1000bp\t598\n",
      "Query Gaps larger than 1000bp\t550\n",
      "Ref N to query bp\t337994\n",
      "Shared seq bp\t75807242\n",
      "Aligned Variance in bp\t1004791\n",
      "Query unique bp\t33191258\n",
      "Ref unique bp\t6846475\n",
      "Query N to ref in bp\t853093\n",
      "Total alignment Length\t76812033\n",
      "Ref Chr Total Size (No N's)\t83658508\n",
      "Ref Identity within Alignment\t0.9869188334072605\n",
      "Alignment Coverage of Ref Chr\t0.9181616411327823\n",
      "Alignment Coverage of Query Main Chr\t0.6982703181125736\n",
      "\n",
      "\n",
      "chr18\n",
      "Ref Number of Gaps (all)\t55079\n",
      "Query Number of Gaps (all)\t56834\n",
      "Query Gaps larger than 10bp\t6424\n",
      "Ref Gaps larger than 10bp\t5861\n",
      "Query Gaps larger than 100bp\t1148\n",
      "Ref Gaps larger than 100bp\t1129\n",
      "Query Gaps larger than 1000bp\t283\n",
      "Ref Gaps larger than 1000bp\t323\n",
      "Ref N to query bp\t283680\n",
      "Ref unique bp\t6855876\n",
      "Shared seq bp\t72457276\n",
      "Aligned Variance in bp\t978685\n",
      "Query unique bp\t1233057\n",
      "Query N to ref in bp\t421540\n",
      "Total alignment Length\t73435961\n",
      "Ref Chr Total Size (No N's)\t80291837\n",
      "Ref Identity within Alignment\t0.9866729462422368\n",
      "Alignment Coverage of Ref Chr\t0.9146130384337825\n",
      "Alignment Coverage of Query Main Chr\t0.9834863637820976\n",
      "\n",
      "\n",
      "chr19\n",
      "Ref Number of Gaps (all)\t53722\n",
      "Ref Gaps larger than 10bp\t7683\n",
      "Query Number of Gaps (all)\t54235\n",
      "Query Gaps larger than 10bp\t8220\n",
      "Query Gaps larger than 100bp\t1978\n",
      "Query Gaps larger than 1000bp\t608\n",
      "Ref Gaps larger than 100bp\t1990\n",
      "Ref Gaps larger than 1000bp\t667\n",
      "Ref N to query bp\t176858\n",
      "Shared seq bp\t52358698\n",
      "Aligned Variance in bp\t831326\n",
      "Ref unique bp\t5993312\n",
      "Query unique bp\t2981850\n",
      "Query N to ref in bp\t509942\n",
      "Total alignment Length\t53190024\n",
      "Ref Chr Total Size (No N's)\t59183336\n",
      "Ref Identity within Alignment\t0.9843706406299045\n",
      "Alignment Coverage of Ref Chr\t0.898733116362349\n",
      "Alignment Coverage of Query Main Chr\t0.946915604061919\n",
      "\n",
      "\n",
      "chr2\n",
      "Ref Number of Gaps (all)\t170672\n",
      "Query Number of Gaps (all)\t178311\n",
      "Query Gaps larger than 10bp\t20429\n",
      "Ref Gaps larger than 10bp\t18015\n",
      "Ref Gaps larger than 100bp\t3618\n",
      "Query Gaps larger than 100bp\t3701\n",
      "Query Gaps larger than 1000bp\t1068\n",
      "Ref Gaps larger than 1000bp\t1187\n",
      "Ref N to query bp\t1645301\n",
      "Ref unique bp\t8260400\n",
      "Shared seq bp\t230534885\n",
      "Aligned Variance in bp\t3039716\n",
      "Query unique bp\t4682573\n",
      "Query N to ref in bp\t512888\n",
      "Total alignment Length\t233574601\n",
      "Ref Chr Total Size (No N's)\t241835001\n",
      "Ref Identity within Alignment\t0.9869861021404464\n",
      "Alignment Coverage of Ref Chr\t0.9658428268619397\n",
      "Alignment Coverage of Query Main Chr\t0.9803465603096594\n",
      "\n",
      "\n",
      "chr20\n",
      "Query Number of Gaps (all)\t47098\n",
      "Ref Number of Gaps (all)\t45779\n",
      "Query Gaps larger than 10bp\t6115\n",
      "Ref Gaps larger than 10bp\t5898\n",
      "Ref Gaps larger than 100bp\t1291\n",
      "Query Gaps larger than 100bp\t1104\n",
      "Ref Gaps larger than 1000bp\t405\n",
      "Query Gaps larger than 1000bp\t335\n",
      "Ref N to query bp\t500575\n",
      "Shared seq bp\t58788482\n",
      "Aligned Variance in bp\t834315\n",
      "Query unique bp\t1090382\n",
      "Ref unique bp\t4674499\n",
      "Query N to ref in bp\t65675\n",
      "Total alignment Length\t59622797\n",
      "Ref Chr Total Size (No N's)\t64297296\n",
      "Ref Identity within Alignment\t0.9860067785816892\n",
      "Alignment Coverage of Ref Chr\t0.9272986689829071\n",
      "Alignment Coverage of Query Main Chr\t0.9820404396877324\n",
      "\n",
      "\n",
      "chr21\n",
      "Query Number of Gaps (all)\t35036\n",
      "Ref Number of Gaps (all)\t33180\n",
      "Ref Gaps larger than 10bp\t4281\n",
      "Query Gaps larger than 10bp\t4377\n",
      "Query Gaps larger than 100bp\t909\n",
      "Query Gaps larger than 1000bp\t323\n",
      "Ref Gaps larger than 100bp\t1027\n",
      "Ref Gaps larger than 1000bp\t326\n",
      "Ref N to query bp\t6621364\n",
      "Shared seq bp\t35871105\n",
      "Aligned Variance in bp\t588978\n",
      "Query unique bp\t910924\n",
      "Ref unique bp\t3798511\n",
      "Query N to ref in bp\t180762\n",
      "Total alignment Length\t36460083\n",
      "Ref Chr Total Size (No N's)\t40258594\n",
      "Ref Identity within Alignment\t0.9838459501038437\n",
      "Alignment Coverage of Ref Chr\t0.9056472016881663\n",
      "Alignment Coverage of Query Main Chr\t0.9756248473582743\n",
      "\n",
      "\n",
      "chr22\n",
      "Query Number of Gaps (all)\t35142\n",
      "Ref Number of Gaps (all)\t33888\n",
      "Query Gaps larger than 10bp\t5098\n",
      "Query Gaps larger than 100bp\t1200\n",
      "Query Gaps larger than 1000bp\t443\n",
      "Ref Gaps larger than 10bp\t5000\n",
      "Ref Gaps larger than 100bp\t1311\n",
      "Ref Gaps larger than 1000bp\t481\n",
      "Ref N to query bp\t11659387\n",
      "Shared seq bp\t35177263\n",
      "Aligned Variance in bp\t598637\n",
      "Ref unique bp\t3699846\n",
      "Query unique bp\t5071519\n",
      "Query N to ref in bp\t278066\n",
      "Total alignment Length\t35775900\n",
      "Ref Chr Total Size (No N's)\t39475746\n",
      "Ref Identity within Alignment\t0.983267031716882\n",
      "Alignment Coverage of Ref Chr\t0.9062754634199947\n",
      "Alignment Coverage of Query Main Chr\t0.8758423635040442\n",
      "\n",
      "\n",
      "chr3\n",
      "Ref Number of Gaps (all)\t137201\n",
      "Query Number of Gaps (all)\t142805\n",
      "Query Gaps larger than 10bp\t15560\n",
      "Ref Gaps larger than 10bp\t13383\n",
      "Ref Gaps larger than 100bp\t2374\n",
      "Query Gaps larger than 100bp\t2525\n",
      "Query Gaps larger than 1000bp\t662\n",
      "Ref Gaps larger than 1000bp\t718\n",
      "Ref N to query bp\t195424\n",
      "Ref unique bp\t6981350\n",
      "Shared seq bp\t189497504\n",
      "Aligned Variance in bp\t2455454\n",
      "Query unique bp\t2722691\n",
      "Query N to ref in bp\t454701\n",
      "Total alignment Length\t191952958\n",
      "Ref Chr Total Size (No N's)\t198934308\n",
      "Ref Identity within Alignment\t0.9872080429205994\n",
      "Alignment Coverage of Ref Chr\t0.9649062543802148\n",
      "Alignment Coverage of Query Main Chr\t0.9860142189637698\n",
      "\n",
      "\n",
      "chr4\n",
      "Query Number of Gaps (all)\t145896\n",
      "Ref Number of Gaps (all)\t139495\n",
      "Query Gaps larger than 10bp\t15261\n",
      "Ref Gaps larger than 10bp\t13788\n",
      "Ref Gaps larger than 100bp\t2773\n",
      "Query Gaps larger than 100bp\t2840\n",
      "Ref Gaps larger than 1000bp\t912\n",
      "Query Gaps larger than 1000bp\t832\n",
      "Ref N to query bp\t461888\n",
      "Ref unique bp\t6721519\n",
      "Shared seq bp\t181435990\n",
      "Aligned Variance in bp\t2501290\n",
      "Query unique bp\t3241496\n",
      "Query N to ref in bp\t880917\n",
      "Total alignment Length\t183937280\n",
      "Ref Chr Total Size (No N's)\t190658799\n",
      "Ref Identity within Alignment\t0.9864013972588918\n",
      "Alignment Coverage of Ref Chr\t0.9647458232441714\n",
      "Alignment Coverage of Query Main Chr\t0.9826823528325669\n",
      "\n",
      "\n",
      "chr5\n",
      "Ref Number of Gaps (all)\t127304\n",
      "Query Number of Gaps (all)\t131709\n",
      "Query Gaps larger than 10bp\t14357\n",
      "Ref Gaps larger than 10bp\t12607\n",
      "Ref Gaps larger than 100bp\t2396\n",
      "Ref Gaps larger than 1000bp\t828\n",
      "Query Gaps larger than 100bp\t2518\n",
      "Query Gaps larger than 1000bp\t725\n",
      "Ref N to query bp\t272881\n",
      "Ref unique bp\t8377245\n",
      "Shared seq bp\t171933266\n",
      "Aligned Variance in bp\t2254345\n",
      "Query unique bp\t30017302\n",
      "Query N to ref in bp\t712412\n",
      "Total alignment Length\t174187611\n",
      "Ref Chr Total Size (No N's)\t182564856\n",
      "Ref Identity within Alignment\t0.9870579486849957\n",
      "Alignment Coverage of Ref Chr\t0.9541135945682777\n",
      "Alignment Coverage of Query Main Chr\t0.853004016607573\n",
      "\n",
      "\n",
      "chr6\n",
      "Ref Number of Gaps (all)\t123905\n",
      "Ref Gaps larger than 10bp\t12484\n",
      "Query Number of Gaps (all)\t125787\n",
      "Ref Gaps larger than 100bp\t2308\n",
      "Ref Gaps larger than 1000bp\t688\n",
      "Query Gaps larger than 10bp\t13711\n",
      "Query Gaps larger than 100bp\t2331\n",
      "Query Gaps larger than 1000bp\t633\n",
      "Ref N to query bp\t727457\n",
      "Shared seq bp\t163553827\n",
      "Aligned Variance in bp\t2128352\n",
      "Ref unique bp\t5024550\n",
      "Query unique bp\t3922784\n",
      "Query N to ref in bp\t424267\n",
      "Total alignment Length\t165682179\n",
      "Ref Chr Total Size (No N's)\t170706729\n",
      "Ref Identity within Alignment\t0.9871540076739334\n",
      "Alignment Coverage of Ref Chr\t0.9705661866439957\n",
      "Alignment Coverage of Query Main Chr\t0.9768710541801775\n",
      "\n",
      "\n",
      "chr7\n",
      "Query Number of Gaps (all)\t124269\n",
      "Ref Number of Gaps (all)\t119253\n",
      "Ref Gaps larger than 10bp\t14143\n",
      "Query Gaps larger than 10bp\t15879\n",
      "Query Gaps larger than 100bp\t3533\n",
      "Query Gaps larger than 1000bp\t1163\n",
      "Ref Gaps larger than 100bp\t3573\n",
      "Ref Gaps larger than 1000bp\t1277\n",
      "Ref N to query bp\t376941\n",
      "Ref unique bp\t8153166\n",
      "Shared seq bp\t149955018\n",
      "Aligned Variance in bp\t2064825\n",
      "Query unique bp\t6402326\n",
      "Query N to ref in bp\t2041529\n",
      "Total alignment Length\t152019843\n",
      "Ref Chr Total Size (No N's)\t160173009\n",
      "Ref Identity within Alignment\t0.9864173981550554\n",
      "Alignment Coverage of Ref Chr\t0.9490977534173688\n",
      "Alignment Coverage of Query Main Chr\t0.9595869312962128\n",
      "\n",
      "\n",
      "chr8\n",
      "Query Number of Gaps (all)\t107404\n",
      "Ref Number of Gaps (all)\t103879\n",
      "Query Gaps larger than 10bp\t12314\n",
      "Query Gaps larger than 100bp\t2183\n",
      "Ref Gaps larger than 10bp\t11382\n",
      "Query Gaps larger than 1000bp\t565\n",
      "Ref Gaps larger than 100bp\t2359\n",
      "Ref Gaps larger than 1000bp\t739\n",
      "Ref N to query bp\t370500\n",
      "Ref unique bp\t6336318\n",
      "Shared seq bp\t137307451\n",
      "Aligned Variance in bp\t1984435\n",
      "Query unique bp\t2323770\n",
      "Query N to ref in bp\t9448\n",
      "Total alignment Length\t139291886\n",
      "Ref Chr Total Size (No N's)\t145628204\n",
      "Ref Identity within Alignment\t0.9857534056219183\n",
      "Alignment Coverage of Ref Chr\t0.956489760733436\n",
      "Alignment Coverage of Query Main Chr\t0.9835910091748613\n",
      "\n",
      "\n",
      "chr9\n",
      "Query Number of Gaps (all)\t95198\n",
      "Ref Number of Gaps (all)\t90319\n",
      "Ref Gaps larger than 10bp\t10275\n",
      "Query Gaps larger than 10bp\t10949\n",
      "Ref Gaps larger than 100bp\t2207\n",
      "Query Gaps larger than 100bp\t2123\n",
      "Query Gaps larger than 1000bp\t713\n",
      "Ref Gaps larger than 1000bp\t845\n",
      "Ref N to query bp\t16604167\n",
      "Ref unique bp\t6501557\n",
      "Shared seq bp\t114437369\n",
      "Aligned Variance in bp\t1639734\n",
      "Query unique bp\t2009904\n",
      "Query N to ref in bp\t240403\n",
      "Total alignment Length\t116077103\n",
      "Ref Chr Total Size (No N's)\t122578660\n",
      "Ref Identity within Alignment\t0.9858737515184196\n",
      "Alignment Coverage of Ref Chr\t0.9469601234015774\n",
      "Alignment Coverage of Query Main Chr\t0.9829794653022241\n",
      "\n",
      "\n",
      "chrX\n",
      "Query Number of Gaps (all)\t97074\n",
      "Ref Number of Gaps (all)\t89016\n",
      "Ref Gaps larger than 10bp\t10685\n",
      "Query Gaps larger than 10bp\t11746\n",
      "Ref Gaps larger than 100bp\t2703\n",
      "Ref Gaps larger than 1000bp\t1077\n",
      "Query Gaps larger than 100bp\t2534\n",
      "Query Gaps larger than 1000bp\t950\n",
      "Ref N to query bp\t1147971\n",
      "Ref unique bp\t10676764\n",
      "Shared seq bp\t143853741\n",
      "Aligned Variance in bp\t1536778\n",
      "Query unique bp\t2966721\n",
      "Query N to ref in bp\t1094639\n",
      "Total alignment Length\t145390519\n",
      "Ref Chr Total Size (No N's)\t156067283\n",
      "Ref Identity within Alignment\t0.9894299985269328\n",
      "Alignment Coverage of Ref Chr\t0.9315887110048555\n",
      "Alignment Coverage of Query Main Chr\t0.9800028566182547\n",
      "\n",
      "\n",
      "chrY\n",
      "Query Number of Gaps (all)\t35425\n",
      "Ref Number of Gaps (all)\t35422\n",
      "Ref Gaps larger than 10bp\t7320\n",
      "Query Gaps larger than 10bp\t7327\n",
      "Query Gaps larger than 100bp\t3123\n",
      "Ref Gaps larger than 100bp\t3226\n",
      "Ref Gaps larger than 1000bp\t1430\n",
      "Query Gaps larger than 1000bp\t1371\n",
      "Ref N to query bp\t30813205\n",
      "Ref unique bp\t4713798\n",
      "Shared seq bp\t22309259\n",
      "Aligned Variance in bp\t758461\n",
      "Query unique bp\t5105610\n",
      "Total alignment Length\t23067720\n",
      "Ref Chr Total Size (No N's)\t27781518\n",
      "Ref Identity within Alignment\t0.9671202442200616\n",
      "Alignment Coverage of Ref Chr\t0.8303261182488301\n",
      "Alignment Coverage of Query Main Chr\t0.8187786108351409\n"
     ]
    }
   ],
   "source": [
    "root = r'E:\\Projects\\FluentDNA-2.4.1\\www-data\\dnadata'\n",
    "# folder = os.path.join(root, )\n",
    "genome = collect_stats(root, prefix='Human Hg38 vs Chimpanzee PanTro6_')\n",
    "for name, stats in genome.items():\n",
    "    print(f\"\\n\\n{name[33:]}\")\n",
    "    for k,v in stats.items():\n",
    "        print(f'{k}\\t{v}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Ref Number of Gaps (all)', 'Ref Gaps larger than 10bp', 'Query Number of Gaps (all)', 'Query Gaps larger than 10bp', 'Ref Gaps larger than 100bp', 'Query Gaps larger than 100bp', 'Ref Gaps larger than 1000bp', 'Query Gaps larger than 1000bp', 'Ref N to query bp', 'Ref unique bp', 'Shared seq bp', 'Aligned Variance in bp', 'Query unique bp', 'Query N to ref in bp', 'Total alignment Length', \"Ref Chr Total Size (No N's)\", 'Ref Identity within Alignment', 'Alignment Coverage of Ref Chr', 'Alignment Coverage of Query Main Chr']\n",
      "chr1\tchr2\tchr3\tchr4\tchr5\tchr6\tchr7\tchr8\tchr9\tchr10\tchr11\tchr12\tchr13\tchr14\tchr15\tchr16\tchr17\tchr18\tchr19\tchr20\tchr21\tchr22\tchrX\tchrY\n",
      "Ref Number of Gaps (all)\t165609\t170672\t137201\t139495\t127304\t123905\t119253\t103879\t90319\t95484\t92132\t96419\t71882\t65255\t63744\t66458\t64312\t55079\t53722\t45779\t33180\t33888\t89016\t35422\t\n",
      "Ref Gaps larger than 10bp\t18763\t18015\t13383\t13788\t12607\t12484\t14143\t11382\t10275\t11050\t9946\t9907\t7416\t6801\t6843\t9996\t8164\t5861\t7683\t5898\t4281\t5000\t10685\t7320\t\n",
      "Query Number of Gaps (all)\t171053\t178311\t142805\t145896\t131709\t125787\t124269\t107404\t95198\t98327\t94230\t98479\t74740\t66865\t66261\t70208\t64542\t56834\t54235\t47098\t35036\t35142\t97074\t35425\t\n",
      "Query Gaps larger than 10bp\t21270\t20429\t15560\t15261\t14357\t13711\t15879\t12314\t10949\t12245\t10997\t11165\t8273\t7557\t7551\t10578\t8707\t6424\t8220\t6115\t4377\t5098\t11746\t7327\t\n",
      "Ref Gaps larger than 100bp\t4479\t3618\t2374\t2773\t2396\t2308\t3573\t2359\t2207\t2504\t1971\t1901\t1638\t1370\t1455\t2892\t1826\t1129\t1990\t1291\t1027\t1311\t2703\t3226\t\n",
      "Query Gaps larger than 100bp\t4832\t3701\t2525\t2840\t2518\t2331\t3533\t2183\t2123\t2354\t1996\t1887\t1586\t1328\t1404\t2723\t1727\t1148\t1978\t1104\t909\t1200\t2534\t3123\t\n",
      "Ref Gaps larger than 1000bp\t1655\t1187\t718\t912\t828\t688\t1277\t739\t845\t819\t677\t586\t506\t462\t578\t1127\t598\t323\t667\t405\t326\t481\t1077\t1430\t\n",
      "Query Gaps larger than 1000bp\t1943\t1068\t662\t832\t725\t633\t1163\t565\t713\t620\t598\t465\t408\t440\t478\t969\t550\t283\t608\t335\t323\t443\t950\t1371\t\n",
      "Ref N to query bp\t18475471\t1645301\t195424\t461888\t272881\t727457\t376941\t370500\t16604167\t534460\t552880\t137493\t16381203\t16475569\t17349864\t8533476\t337994\t283680\t176858\t500575\t6621364\t11659387\t1147971\t30813205\t\n",
      "Ref unique bp\t8720120\t8260400\t6981350\t6721519\t8377245\t5024550\t8153166\t6336318\t6501557\t6082908\t7284509\t6103971\t3849601\t4245909\t4447223\t4824479\t6846475\t6855876\t5993312\t4674499\t3798511\t3699846\t10676764\t4713798\t\n",
      "Shared seq bp\t220917707\t230534885\t189497504\t181435990\t171933266\t163553827\t149955018\t137307451\t114437369\t126144862\t126357568\t126059664\t93347690\t85325045\t79377739\t76802346\t75807242\t72457276\t52358698\t58788482\t35871105\t35177263\t143853741\t22309259\t\n",
      "Aligned Variance in bp\t2830836\t3039716\t2455454\t2501290\t2254345\t2128352\t2064825\t1984435\t1639734\t1708138\t1694144\t1618815\t1250938\t1134250\t1110620\t1219533\t1004791\t978685\t831326\t834315\t588978\t598637\t1536778\t758461\t\n",
      "Query unique bp\t8724657\t4682573\t2722691\t3241496\t30017302\t3922784\t6402326\t2323770\t2009904\t4145888\t2765835\t1956826\t5205275\t2098847\t9776198\t8134807\t33191258\t1233057\t2981850\t1090382\t910924\t5071519\t2966721\t5105610\t\n",
      "Query N to ref in bp\t2495044\t512888\t454701\t880917\t712412\t424267\t2041529\t9448\t240403\t1197562\t8258\t5774\t402562\t475733\t1518735\t355075\t853093\t421540\t509942\t65675\t180762\t278066\t1094639\t0\t\n",
      "Total alignment Length\t223748543\t233574601\t191952958\t183937280\t174187611\t165682179\t152019843\t139291886\t116077103\t127853000\t128051712\t127678479\t94598628\t86459295\t80488359\t78021879\t76812033\t73435961\t53190024\t59622797\t36460083\t35775900\t145390519\t23067720\t\n",
      "Ref Chr Total Size (No N's)\t232468663\t241835001\t198934308\t190658799\t182564856\t170706729\t160173009\t145628204\t122578660\t133935908\t135336221\t133782450\t98448229\t90705204\t84935582\t82846358\t83658508\t80291837\t59183336\t64297296\t40258594\t39475746\t156067283\t27781518\t\n",
      "Ref Identity within Alignment\t0.9873481366088717\t0.9869861021404464\t0.9872080429205994\t0.9864013972588918\t0.9870579486849957\t0.9871540076739334\t0.9864173981550554\t0.9857534056219183\t0.9858737515184196\t0.986639828553104\t0.986769844982627\t0.987321160052353\t0.9867763621265205\t0.9868811097754151\t0.9862014828753062\t0.9843693459369263\t0.9869188334072605\t0.9866729462422368\t0.9843706406299045\t0.9860067785816892\t0.9838459501038437\t0.983267031716882\t0.9894299985269328\t0.9671202442200616\t\n",
      "Alignment Coverage of Ref Chr\t0.9624890516964001\t0.9658428268619397\t0.9649062543802148\t0.9647458232441714\t0.9541135945682777\t0.9705661866439957\t0.9490977534173688\t0.956489760733436\t0.9469601234015774\t0.9545834415069632\t0.9461747273111756\t0.9543739033034602\t0.9608972041538706\t0.9531900176311825\t0.9476400479601117\t0.9417659494458405\t0.9181616411327823\t0.9146130384337825\t0.898733116362349\t0.9272986689829071\t0.9056472016881663\t0.9062754634199947\t0.9315887110048555\t0.8303261182488301\t\n",
      "Alignment Coverage of Query Main Chr\t0.9624702675405165\t0.9803465603096594\t0.9860142189637698\t0.9826823528325669\t0.853004016607573\t0.9768710541801775\t0.9595869312962128\t0.9835910091748613\t0.9829794653022241\t0.9685914929828803\t0.9788573087981844\t0.9849051460171286\t0.9478449755617273\t0.9762997850609829\t0.8916939458308093\t0.9055812453139156\t0.6982703181125736\t0.9834863637820976\t0.946915604061919\t0.9820404396877324\t0.9756248473582743\t0.8758423635040442\t0.9800028566182547\t0.8187786108351409\t\n"
     ]
    }
   ],
   "source": [
    "chr_order = 'chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 chrX chrY'.split()\n",
    "prefix = 'Human Hg38 vs Chimpanzee PanTro6_'\n",
    "labels = list(genome['Human Hg38 vs Chimpanzee PanTro6_chr1'].keys())\n",
    "print(labels)\n",
    "print('\\t'.join(chr_order))\n",
    "for label in labels:\n",
    "    if label:\n",
    "        print(label, end='\\t')\n",
    "        for c in chr_order:\n",
    "            \n",
    "            v = genome[prefix + c][label] if label in genome[prefix + c] else 0\n",
    "            print(v, end='\\t')\n",
    "        print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hg38toPantro6 identity 98.67%\n",
      "hg38toPantro6 coverage 91.46%\n",
      "\n"
     ]
    }
   ],
   "source": [
    "all_results = {}\n",
    "for comparison in ['hg38toPantro6']:\n",
    "    all_results[comparison] = collect_stats(comparison)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate centromere sizes\n",
    "Parse a CSV with chromosome name, and two positions.  Output the distance between the positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3370400\t2076300\t3157700\t2077200\t4186600\t1549700\t4162600\t2802300\t2600400\t2862200\t4463500\t3252100\t2049700\t2306200\t2779400\t2481000\t4171000\t5434500\t3408800\t3163400\t2244000\t2516800\t4574000\t662700\t"
     ]
    }
   ],
   "source": [
    "centromeres = {}\n",
    "lines = open(r\"Unique Human Centromere Locations.txt\", 'r').readlines()\n",
    "for line_i in range(0, len(lines), 2):\n",
    "    label1, start = lines[line_i].split('\\t')\n",
    "    label2, end = lines[line_i+1].split('\\t')\n",
    "    start, end = int(start), int(end)\n",
    "    assert label1 == label2 and start < end, (label1, label2, start, end)\n",
    "    centromeres[label1] = end - start\n",
    "for c in chr_order:\n",
    "    print(centromeres[c], end='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
